{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Idea\n",
    "\n",
    "The idea here is to trim the sequence up to the cut on both reads and then do a global alignment end-to-end agasint the human genome.\n",
    "\n",
    "The idea here is to take the files that are 'trimmed' and 'correct priming' checking. Then to do another round of trimming to get down to the cut site. I need to make another folder for 'trimmed_to_cut'\n",
    "\n",
    "For Tn5 setup the 'Read 1' contains the sequence away from the gene specific primer. So it is the read we should concentration on if we don't do paired end.\n",
    "\n",
    "For LAM the 'Read 2' is the the sequence distal to the gene specific primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "\n",
    "#not sure if I need these\n",
    "class Error(Exception):\n",
    "    \"\"\"Base class for exceptions in this module.\"\"\"\n",
    "    pass\n",
    "\n",
    "class StrandError(Error):\n",
    "    \"\"\"Exception raised for errors in the strand information.\n",
    "    Attributes:\n",
    "        expression -- input expression in which the error occurred\n",
    "        message -- explanation of the error\n",
    "    \"\"\"\n",
    "    def __init__(self, message):\n",
    "        self.message = message\n",
    "\n",
    "class ReactionTypeError(Error):\n",
    "    \"\"\"Exception raised for errors in the reaction type to be processed.\n",
    "    Attributes:\n",
    "        expression -- input expression in which the error occurred\n",
    "        message -- explanation of the error\n",
    "    \"\"\"\n",
    "    def __init__(self, message):\n",
    "        self.message = message\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inputs\n",
    "#### next few window are different inputs to make it run well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#things taken from original uditas args\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this is for polb fwd. row 0 in sheet\n"
     ]
    }
   ],
   "source": [
    "#This is for the sequence to the breaksite for the pipeline of genomic alignment \n",
    "\n",
    "print('this is for polb fwd. row 0 in sheet')\n",
    "#seq_primer_to_breaksite = 'TGGTTTTATTTCACCCCATGATAGTAATTATATCACTTCTGATCTGTTAAGAATAGACCTTTTAAAAGTATTGGATAACTTAGAGATGAGACATCTTCAGTTACTCTGTTATTCACCTATTACTCCTTAGGTTACTTGTGAATAATTTTGTGTGGGTCA'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_BCL2Fastq_only\n"
     ]
    }
   ],
   "source": [
    "#Directory\n",
    "\n",
    "directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_BCL2Fastq_only'\n",
    "\n",
    "#this is a minimal directory of only 300 files for quick debugging\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_BCL2Fastq_only/10000Reads'\n",
    "\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                        Tn5_Pol50_pbF\n",
       "Sample                                                                               Polb\n",
       "description                                   MiniSeq_K562_PolbHIROS_50per_unsorted_polbF\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                      prE368\n",
       "I7_Index_ID                                                          P7_N701_SBS12nextera\n",
       "index_I1                                                                             N701\n",
       "barcode_I1                                                                       TAAGGCGA\n",
       "I5_Index_ID                                                              P5_UMI_N50_Tn5-A\n",
       "index_I2                                                                             N501\n",
       "barcode_I2                                                                       GCGATCTA\n",
       "primer1_ForNest                                                                     prE86\n",
       "primer                                                                             prE361\n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                         4.23497e+07\n",
       "end                                                                           4.23498e+07\n",
       "strand                                                                                  +\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                                         pE049\n",
       "plasmid_sequence                        ctatggaaaaacgccagcaacgcggcctttttacggttcctggcct...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Function to make the 'amplicon_info' list. Taking the line of our experiment csv file\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "\n",
    "#the amplicon info is related to the line on the csv file. It is indexed from 0. For PolbF we use 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "#0 is because this polbFwd is in the first line. Need to change.\n",
    "amplicon_info = get_csv_data(directory, 0)\n",
    "\n",
    "#1 is because this polbRev \n",
    "#amplicon_info = get_csv_data(directory, 1)\n",
    "\n",
    "#2 is because this mcherry_rev\n",
    "#amplicon_info = get_csv_data(directory, 2)\n",
    "\n",
    "#3 is because this mcherry_fwd\n",
    "#amplicon_info = get_csv_data(directory, 3)\n",
    "\n",
    "\n",
    "###########check waht is printed below #######\n",
    "amplicon_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "assembly = amplicon_info['genome']\n",
    "file_genome_2bit = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly + '.2bit')\n",
    "print(file_genome_2bit)\n",
    "\n",
    "# BOWTIE2_INDEXES are needed for global alignments\n",
    "#not sure if this will work\n",
    "#normally in bash: export BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "\n",
    "#Needs to have a bowtie build in which the targeting vector is included!!\n",
    "\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "\n",
    "#check the env variables\n",
    "#%env"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions to process the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#these are copied and unchanged from the Uditas v1 software\n",
    "\n",
    "################################################################################\n",
    "# Open .fastq or .fastq.gz files for reading\n",
    "################################################################################\n",
    "def open_fastq_or_gz(filename):\n",
    "    if filename.endswith(\".fastq\") and os.access(filename, os.F_OK):\n",
    "        return open(filename, \"rU\")\n",
    "    elif filename.endswith(\".fastq.gz\") and os.access(filename, os.F_OK):\n",
    "        return gzip.open(filename, \"rb\")\n",
    "    elif filename.endswith(\".fastq\") and os.access(filename + \".gz\", os.F_OK):\n",
    "        return gzip.open(filename + \".gz\", \"rb\")\n",
    "    elif filename.endswith(\".fastq.gz\") and os.access(filename[:-3], os.F_OK):\n",
    "        return open(filename[:-3], \"rU\")\n",
    "    raise IOError(\"Unknown file: \" + filename)\n",
    "\n",
    "################################################################################\n",
    "# Hamming distance\n",
    "# From http://code.activestate.com/recipes/499304-hamming-distance/\n",
    "################################################################################\n",
    "def hamm_dist(str1, str2):\n",
    "    assert len(str1) == len(str2)\n",
    "    ne = operator.ne\n",
    "    return sum(itertools.imap(ne, str1, str2))\n",
    "\n",
    "################################################################################\n",
    "# Select closest barcode with a maximum number of mismatches\n",
    "# By default it returns barcodes with a maximum of n_max_mismatches=2 mismatches\n",
    "################################################################################\n",
    "def select_barcode(seq, barcode_list, n_max_mismatches=1):\n",
    "    # This compares with all barcodes and selects the one with the smallest hamming distance\n",
    "    # Before calling this function check if the sequence is already a barcode\n",
    "    matched_barcodes = list()\n",
    "    distances = list()\n",
    "    for barcode in barcode_list:\n",
    "        h_d = hamm_dist(seq, barcode)\n",
    "        if h_d <= n_max_mismatches:\n",
    "            matched_barcodes.append(barcode)\n",
    "            distances.append(h_d)\n",
    "    indices = [i for i, x in enumerate(distances) if x == min(distances)]\n",
    "    return [matched_barcodes[i] for i in indices]\n",
    "\n",
    "\n",
    "################################################################################\n",
    "# Mask sequence by quality score\n",
    "################################################################################\n",
    "def mask(seq, qual, min_qual=12):\n",
    "\n",
    "    return \"\".join((b if (ord(q) - 33) >= min_qual else \"N\") for b, q in itertools.izip(seq, qual))\n",
    "\n",
    "\n",
    "################################################################################\n",
    "# get the reverse-complement DNA sequence\n",
    "################################################################################\n",
    "def reverse_complement(seq):\n",
    "    seq_dict = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G', 'N': 'N', 'a': 't', 't': 'a', 'g': 'c', 'c': 'g'}\n",
    "    return \"\".join([seq_dict[base] for base in reversed(seq)])\n",
    "\n",
    "\n",
    "################################################################################\n",
    "# Create umi dict\n",
    "################################################################################\n",
    "def create_umi_dict(filename):\n",
    "\n",
    "    umi_file = open_fastq_or_gz(filename)\n",
    "\n",
    "    umi_dict = dict()\n",
    "\n",
    "    umi_reads = itertools.izip(umi_file)\n",
    "\n",
    "    for header_umi in umi_reads:\n",
    "\n",
    "        seq_umi = umi_reads.next()\n",
    "        umi_reads.next()\n",
    "        qual_umi = umi_reads.next()\n",
    "        umi_dict[header_umi[0].split()[0][1:]] = [seq_umi[0].rstrip(), qual_umi[0].rstrip()]\n",
    "\n",
    "    return umi_dict\n",
    "\n",
    "\n",
    "\n",
    "################################################################################\n",
    "# create list of output files\n",
    "#I added a bit to allow for the function that makes a fastq of the correctly primed targets\n",
    "#I  need to add to make  mishas umi file\n",
    "################################################################################\n",
    "def create_filename(dir_sample, N7, N5, filetype):\n",
    "    main_folder = os.path.join(dir_sample, N7 + '_' + N5)\n",
    "    if filetype == 'mainfolder':\n",
    "        return main_folder\n",
    "    elif filetype == 'amplicons':\n",
    "        return os.path.join(main_folder, 'amplicons')\n",
    "    elif filetype == 'R1fastq':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.fastq')\n",
    "    elif filetype == 'R1fastqgz':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.fastq.gz')\n",
    "    elif filetype == 'R2fastq':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.fastq')\n",
    "    elif filetype == 'R2fastqgz':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.fastq.gz')\n",
    "   \n",
    "    ### Added to make LAM cleanup after extracting UMI and cleaving adapter\n",
    "    elif filetype == 'R1fastq_LAM':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.LAM.fastq')\n",
    "    elif filetype == 'R1fastqgz_LAM':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.LAM.fastq.gz')\n",
    "    elif filetype == 'R2fastq_LAM':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.LAM.fastq')\n",
    "    elif filetype == 'R2fastqgz_LAM':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.LAM.fastq.gz')\n",
    "    \n",
    "    #### I added these as the sequences that were correctly primed\n",
    "    elif filetype == 'R1fastq_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.CorrPrime.fastq')\n",
    "    elif filetype == 'R1fastqgz_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.CorrPrime.fastq.gz')\n",
    "    elif filetype == 'R2fastq_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.CorrPrime.fastq')\n",
    "    elif filetype == 'R2fastqgz_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.CorrPrime.fastq.gz')\n",
    "    #####\n",
    "    \n",
    "    elif filetype == 'umifastq':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_umi.fastq')\n",
    "    elif filetype == 'umifastqgz':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_umi.fastq.gz')\n",
    "    \n",
    "    ### Added these for trimming to breaksite\n",
    "    elif filetype == 'R1trimmed2break':\n",
    "        return os.path.join(main_folder, 'cutadapt_files_to_breaksite', N7 + '_' + N5 + '_R1.trimmed2break.fastq.gz')\n",
    "    elif filetype == 'R2trimmed2break':\n",
    "        return os.path.join(main_folder, 'cutadapt_files_to_breaksite', N7 + '_' + N5 + '_R2.trimmed2break.fastq.gz')\n",
    "    elif filetype == 'trimmed_report2break':\n",
    "        return os.path.join(main_folder, 'cutadapt_files_to_breaksite', N7 + '_' + N5 + '.trimmed.report.txt')\n",
    "    # added. for aligning to amplicons first\n",
    "    elif filetype == 'break_trimmed_sam_amplicons':         \n",
    "        return os.path.join(main_folder, 'break_trimmed_sam_amplicon_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'break_trimmed_sam_report_amplicons':\n",
    "        return os.path.join(main_folder, 'break_trimmed_sam_amplicon_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'break_trimmed_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_amplicon_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'break_trimmed_sorted_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'break_trimmed_sorted_bai_amplicons':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'break_trimmed_unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_amplicon_files', N7 + '_' + N5 + '_amplicons_unmapped.bam')\n",
    "    elif filetype == 'break_trimmed_qsorted_unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_amplicon_files', N7 + '_' + N5 + '_qsorted_amplicons_unmapped.bam')\n",
    "    elif filetype == 'break_trimmed_unmapped_amplicons_R1fastq':\n",
    "        return os.path.join(main_folder, 'break_trimmed_amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq')\n",
    "    elif filetype == 'break_trimmed_unmapped_amplicons_R2fastq':\n",
    "        return os.path.join(main_folder, 'break_trimmed_amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq')\n",
    "    elif filetype == 'break_trimmed_unmapped_amplicons_R1fastqgz':\n",
    "        return os.path.join(main_folder, 'break_trimmed_amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq.gz')\n",
    "    elif filetype == 'break_trimmed_unmapped_amplicons_R2fastqgz':\n",
    "        return os.path.join(main_folder, 'break_trimmed_amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq.gz')\n",
    "    elif filetype == 'break_trimmed_unmapped_amplicons_report':\n",
    "        return os.path.join(main_folder, 'break_trimmed_amplicons_unmapped_fastq_files', N7 + '_' + N5 + '.unmapped.report.txt')\n",
    "    elif filetype == 'break_trimmed_sam_genome_global':\n",
    "        return os.path.join(main_folder, 'break_trimmed_sam_genome_global_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'break_trimmed_sam_report_genome_global':\n",
    "        return os.path.join(main_folder, 'break_trimmed_sam_genome_global_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'break_trimmed_bam_genome_global':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_genome_global_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'break_trimmed_sorted_bam_genome_global':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'break_trimmed_sorted_bai_genome_global':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'break_trimmed_sam_genome_local':\n",
    "        return os.path.join(main_folder, 'break_trimmed_sam_genome_local_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'break_trimmed_sam_report_genome_local':\n",
    "        return os.path.join(main_folder, 'break_trimmed_sam_genome_local_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'break_trimmed_bam_genome_local':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_genome_local_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'break_trimmed_sorted_bam_genome_local':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'break_trimmed_sorted_bai_genome_local':\n",
    "        return os.path.join(main_folder, 'break_trimmed_bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    ###############################################\n",
    "    \n",
    "    elif filetype == 'R1trimmed':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '_R1.trimmed.fastq.gz')\n",
    "    elif filetype == 'R2trimmed':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '_R2.trimmed.fastq.gz')\n",
    "    elif filetype == 'trimmed_report':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '.trimmed.report.txt')\n",
    "    elif filetype == 'sam_genome_local':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_genome_local':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'sam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'sam_plasmid_local_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_plasmid_local':\n",
    "        return os.path.join(main_folder, 'sam_plasmid_local_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'unmapped_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '_unmapped.bam')\n",
    "    elif filetype == 'qsorted_unmapped_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '_qsorted_unmapped.bam')\n",
    "    elif filetype == 'unmapped_plasmid_R1fastq':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R1.fastq')\n",
    "    elif filetype == 'unmapped_plasmid_R2fastq':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R2.fastq')\n",
    "    elif filetype == 'unmapped_plasmid_R1fastqgz':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R1.fastq.gz')\n",
    "    elif filetype == 'unmapped_plasmid_R2fastqgz':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R2.fastq.gz')\n",
    "    elif filetype == 'sam_amplicons':\n",
    "        return os.path.join(main_folder, 'sam_amplicon_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_amplicons':\n",
    "        return os.path.join(main_folder, 'sam_amplicon_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '_amplicons_unmapped.bam')\n",
    "    elif filetype == 'qsorted_unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '_qsorted_amplicons_unmapped.bam')\n",
    "    elif filetype == 'unmapped_amplicons_R1fastq':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq')\n",
    "    elif filetype == 'unmapped_amplicons_R2fastq':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq')\n",
    "    elif filetype == 'unmapped_amplicons_R1fastqgz':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files',\n",
    "                            N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq.gz')\n",
    "    elif filetype == 'unmapped_amplicons_R2fastqgz':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files',\n",
    "                            N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq.gz')\n",
    "    elif filetype == 'unmapped_amplicons_report':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '.unmapped.report.txt')\n",
    "    elif filetype == 'sam_genome_global':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_genome_global':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'results_amplicons':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5)  # We will append the window size later\n",
    "    elif filetype == 'results_plasmid':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_plasmid.xlsx')\n",
    "    elif filetype == 'results_all_amplicons':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_all_amplicons.xlsx')\n",
    "    elif filetype == 'results_genomewide':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_genomewide.xlsx')\n",
    "    elif filetype == 'summary_all_alignments':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_summary_all_alignments.xlsx')\n",
    "    elif filetype == 'read_counts':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_read_counts.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################\n",
    "#\n",
    "# this was changed from the uditas original. It assumes I already went through the jupyternotebook of the\n",
    "# normal uditias analysis. In short it assumes you already ran the correct_priming() function on the data\n",
    "#\n",
    "# This is used to trim the fastq files. Originally it was just to trim off the adapter sequences.\n",
    "# However in this case we use it to trim off the sequences up to the break\n",
    "\n",
    "# for Tn5 use the reverse compliment and read2. \n",
    "\n",
    "# ##########################\n",
    "def trim_fastq_to_break(dir_sample, amplicon_info, seq_primer_to_breaksite):\n",
    "\n",
    "    \n",
    "    seqRC = reverse_complement(seq_primer_to_breaksite)\n",
    "     \n",
    "    \n",
    "    # We first check if the experiment had any guides\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "\n",
    "    #for Tn5 we only use read 1 as it is the sequence distal to the gene specific primer\n",
    "    file_R1 = create_filename(dir_sample, N7, N5, 'R1fastq_CorrPrime')\n",
    "    #file_R2 = create_filename(dir_sample, N7, N5, 'R2fastq_CorrPrime')\n",
    "    \n",
    "    # trimmedoutput files\n",
    "    file_cutadapt_R1 = create_filename(dir_sample, N7, N5, 'R1trimmed2break')\n",
    "    #file_cutadapt_R2 = create_filename(dir_sample, N7, N5, 'R2trimmed2break')\n",
    "    file_cutadapt_report = create_filename(dir_sample, N7, N5, 'trimmed_report2break')\n",
    "        \n",
    "    if not os.path.exists(os.path.dirname(file_cutadapt_R1)):\n",
    "        os.mkdir(os.path.dirname(file_cutadapt_R1))\n",
    "\n",
    "    # remove adapters with cutadapt\n",
    "    #original uditas peramiter had an error -e 0.33 (but was cutting of random stuff too much)\n",
    "    cutadapt_command = ['cutadapt',\n",
    "                        '-m', '20',\n",
    "                        '-e', '0.20',\n",
    "                        '-a', seqRC,\n",
    "                        '-o', file_cutadapt_R1,\n",
    "                        file_R1]\n",
    "\n",
    "    handle_cutadapt_report = open(file_cutadapt_report, 'wb')\n",
    "    subprocess.call(cutadapt_command, stdout=handle_cutadapt_report)\n",
    "    handle_cutadapt_report.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "trim_fastq_to_break(directory, amplicon_info, seq_primer_to_breaksite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################\n",
    "#\n",
    "# Aligns reads globally to amplicon. \"end-to-end\" as the default function in bowtie2\n",
    "# Input: read trimmed up to the cut site\n",
    "#        align to our reference amplicons to remove reads thats would align there\n",
    "#\n",
    "# ##########################\n",
    "def align_afterbreak_reads_to_amplicons(dir_sample, amplicon_info, ncpu=4):\n",
    "\n",
    "    # We first check if the experiment had any guides\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "\n",
    "    #input file (for Tn5 we use R1 and LAM R2)\n",
    "    file_R1 = create_filename(dir_sample, N7, N5, 'R1trimmed2break')\n",
    "    #file_R2 = create_filename(dir_sample, N7, N5, 'R2trimmed2break')    \n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_R1)):\n",
    "        os.mkdir(os.path.dirname(file_R1))\n",
    "\n",
    "    \n",
    "    file_sam_amplicons = create_filename(dir_sample, N7, N5, 'break_trimmed_sam_amplicons')\n",
    "    file_sam_report_amplicons = create_filename(dir_sample, N7, N5, 'break_trimmed_sam_report_amplicons')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_sam_amplicons)):\n",
    "        os.mkdir(os.path.dirname(file_sam_amplicons))\n",
    "\n",
    "    file_bam_amplicons = create_filename(dir_sample, N7, N5, 'break_trimmed_bam_amplicons')\n",
    "    file_sorted_bam_amplicons = create_filename(dir_sample, N7, N5, 'break_trimmed_sorted_bam_amplicons')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_bam_amplicons)):\n",
    "        os.mkdir(os.path.dirname(file_bam_amplicons))\n",
    "\n",
    "    # global alignment to the amplicons with bowtie2\n",
    "    initial_dir = os.getcwd()\n",
    "    folder_amplicons = create_filename(dir_sample, N7, N5, 'amplicons')\n",
    "\n",
    "    os.chdir(folder_amplicons)\n",
    "    bowtie2_command = ['bowtie2', '-p', str(ncpu), '--very-sensitive',\n",
    "                       '-X', '5000', '-k', '2', '-x', 'amplicons',\n",
    "                       '-U', file_R1, '-S', file_sam_amplicons]\n",
    "\n",
    "    handle_sam_report_amplicons = open(file_sam_report_amplicons, 'wb')\n",
    "\n",
    "    subprocess.call(bowtie2_command, stderr=handle_sam_report_amplicons)\n",
    "\n",
    "    handle_sam_report_amplicons.close()\n",
    "\n",
    "    # convert sam to bam\n",
    "    sam_to_bam_amplicons_command = ['samtools', 'view', '-Sb', file_sam_amplicons]\n",
    "\n",
    "    handle_file_bam_amplicons = open(file_bam_amplicons, 'wb')\n",
    "\n",
    "    subprocess.call(sam_to_bam_amplicons_command, stdout=handle_file_bam_amplicons)\n",
    "\n",
    "    # sort bam files\n",
    "    sort_bam_amplicons_command = ['samtools', 'sort', file_bam_amplicons, '-o', file_sorted_bam_amplicons]\n",
    "\n",
    "    subprocess.call(sort_bam_amplicons_command)\n",
    "\n",
    "    # Clean up\n",
    "    #os.remove(file_sam_amplicons)\n",
    "    os.remove(file_bam_amplicons)\n",
    "\n",
    "    # Create bam index files\n",
    "    create_bam_amplicons_index_command = ['samtools', 'index', file_sorted_bam_amplicons]\n",
    "    subprocess.call(create_bam_amplicons_index_command)\n",
    "\n",
    "    os.chdir(initial_dir)\n",
    "\n",
    "################################################################################\n",
    "# Function to extract all unmapped reads to the amplicons\n",
    "################################################################################\n",
    "def extract_unmapped_afterbreak_to_amplicon_reads(dir_sample, amplicon_info):\n",
    "\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "\n",
    "    file_sorted_bam_amplicons = create_filename(dir_sample, N7, N5, 'break_trimmed_sorted_bam_amplicons')\n",
    "\n",
    "    file_unmapped_bam_amplicons = create_filename(dir_sample, N7, N5, 'break_trimmed_unmapped_bam_amplicons')\n",
    "\n",
    "    file_qsorted_unmapped_bam_amplicons = create_filename(dir_sample, N7, N5, 'break_trimmed_qsorted_unmapped_bam_amplicons')\n",
    "\n",
    "    file_R1_unmapped = create_filename(dir_sample, N7, N5, 'break_trimmed_unmapped_amplicons_R1fastq')\n",
    "    file#_R2_unmapped = create_filename(dir_sample, N7, N5, 'unmapped_amplicons_R2fastq')\n",
    "    file_unmapped_report = create_filename(dir_sample, N7, N5, 'break_trimmed_unmapped_amplicons_report')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_R1_unmapped)):\n",
    "        os.mkdir(os.path.dirname(file_R1_unmapped))\n",
    "\n",
    "    extract_unmapped_bam_command = ['samtools', 'view', '-b', '-f', '0x4', file_sorted_bam_amplicons, '-o',\n",
    "                                    file_unmapped_bam_amplicons]\n",
    "\n",
    "    subprocess.call(extract_unmapped_bam_command)\n",
    "\n",
    "    qsort_unmapped_bam_command = ['samtools', 'sort', '-n', file_unmapped_bam_amplicons, '-o',\n",
    "                                  file_qsorted_unmapped_bam_amplicons]\n",
    "\n",
    "    subprocess.call(qsort_unmapped_bam_command)\n",
    "\n",
    "    bamtofastq_command = ['bedtools', 'bamtofastq', '-i', file_qsorted_unmapped_bam_amplicons,\n",
    "                          '-fq', file_R1_unmapped]\n",
    "\n",
    "    handle_unmapped_report = open(file_unmapped_report, 'wb')\n",
    "    subprocess.call(bamtofastq_command, stderr=handle_unmapped_report)\n",
    "\n",
    "    for fo in [file_R1_unmapped]:\n",
    "        with open(fo) as f_in, gzip.open(fo + '.gz', 'wb') as f_out:\n",
    "            f_out.writelines(f_in)\n",
    "        os.remove(fo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "align_afterbreak_reads_to_amplicons(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "extract_unmapped_afterbreak_to_amplicon_reads(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block (<ipython-input-80-bfa004870f1c>, line 102)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-80-bfa004870f1c>\"\u001b[0;36m, line \u001b[0;32m102\u001b[0m\n\u001b[0;31m    file_sam_genome_local = create_filename(dir_sample, N7, N5, 'break_trimmed_sam_genome_local')\u001b[0m\n\u001b[0m                        ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m expected an indented block\n"
     ]
    }
   ],
   "source": [
    "############################\n",
    "#\n",
    "# Align reads that did not map to the amplicons of interest\n",
    "#\n",
    "# Input: directory to be analyzed\n",
    "#        amplicon_info, slice of sample_info.csv for the sample being processed\n",
    "#        assembly, name of the assembly to be used by bowtie2. It is convenient to place it in a folder especified by\n",
    "#        the environmental variable BOWTIE2_INDEXES\n",
    "#\n",
    "# ##########################\n",
    "def align_afterbreak_genome_global(dir_sample, amplicon_info, assembly, ncpu=4):\n",
    "\n",
    "    # We first check if the experiment had any guides\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    \n",
    "    #input\n",
    "    # Read1 for Tn5 Read2 for LAM\n",
    "    file_R1 = create_filename(dir_sample, N7, N5, 'break_trimmed_unmapped_amplicons_R1fastqgz')\n",
    "    #file_R2 = create_filename(dir_sample, N7, N5, 'unmapped_amplicons_R2fastqgz')\n",
    "\n",
    "    file_sam_genome_global = create_filename(dir_sample, N7, N5, 'break_trimmed_sam_genome_global')\n",
    "    file_sam_report_genome_global = create_filename(dir_sample, N7, N5, 'break_trimmed_sam_report_genome_global')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_sam_genome_global)):\n",
    "        os.mkdir(os.path.dirname(file_sam_genome_global))\n",
    "\n",
    "    file_bam_genome_global = create_filename(dir_sample, N7, N5, 'break_trimmed_bam_genome_global')\n",
    "    file_sorted_bam_genome_global = create_filename(dir_sample, N7, N5, 'break_trimmed_sorted_bam_genome_global')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_bam_genome_global)):\n",
    "        os.mkdir(os.path.dirname(file_bam_genome_global))\n",
    "\n",
    "    # global alignment to the genome with bowtie2\n",
    "    initial_dir = os.getcwd()\n",
    "\n",
    "    bowtie2_command = ['bowtie2', '--very-sensitive', '-p', str(ncpu),\n",
    "                       '-X', '5000', '-k', '2', '-x', assembly,\n",
    "                       '-U', file_R1, '-S', file_sam_genome_global]\n",
    "\n",
    "    handle_sam_report_genome_global = open(file_sam_report_genome_global, 'wb')\n",
    "\n",
    "    subprocess.call(bowtie2_command, stderr=handle_sam_report_genome_global)\n",
    "\n",
    "    handle_sam_report_genome_global.close()\n",
    "\n",
    "    # convert sam to bam\n",
    "    sam_to_bam_genome_global_command = ['samtools', 'view', '-Sb', file_sam_genome_global]\n",
    "\n",
    "    handle_file_bam_genome_global = open(file_bam_genome_global, 'wb')\n",
    "\n",
    "    subprocess.call(sam_to_bam_genome_global_command, stdout=handle_file_bam_genome_global)\n",
    "\n",
    "    # sort bam files\n",
    "    sort_bam_genome_global_command = ['samtools', 'sort', file_bam_genome_global, '-o', file_sorted_bam_genome_global]\n",
    "\n",
    "    subprocess.call(sort_bam_genome_global_command)\n",
    "\n",
    "    # Clean up\n",
    "    #os.remove(file_sam_genome_global)\n",
    "    os.remove(file_bam_genome_global)\n",
    "\n",
    "    # Create bam index files\n",
    "    create_bam_genome_global_index_command = ['samtools', 'index', file_sorted_bam_genome_global]\n",
    "    subprocess.call(create_bam_genome_global_index_command)\n",
    "\n",
    "    os.chdir(initial_dir)\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "############################\n",
    "#       If you put in unmapped = 1 then it uses the unmapped to amplicons read, otherwise it just maps from trimmed fastq\n",
    "# Aligns reads to the whole genome using bowtie2 and local alignment. This is needed to find translocations using\n",
    "#   split reads and a program like socrates\n",
    "#   --local aligment means that the sequence can be soft clipped at ends and is used in translocation programs that then align the clipped ends.\n",
    "#\n",
    "# Input: directory to be analyzed\n",
    "#        amplicon_info, slice of sample_info.csv for the sample being processed\n",
    "#        assembly, name of the assembly to be used by bowtie2. It is convenient to place it in a folder especified by\n",
    "#        the environmental variable BOWTIE2_INDEXES\n",
    "#\n",
    "# ##########################\n",
    "def align_afterbreaks_genome_local(dir_sample, unmapped_amplicons, amplicon_info, assembly):\n",
    "\n",
    "    # We first check if the experiment had any guides\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    \n",
    "    #input\n",
    "    # Read1 for Tn5 Read2 for LAM\n",
    "    \n",
    "    #If we want the trimmed ones\n",
    "    if unmapped_amplicons == 1:\n",
    "        file_R1 = create_filename(dir_sample, N7, N5, 'break_trimmed_unmapped_amplicons_R1fastqgz')\n",
    "        #file_R2 = create_filename(dir_sample, N7, N5, 'unmapped_amplicons_R2fastqgz')\n",
    "\n",
    "    else:\n",
    "        #file_R1 = create_filename(dir_sample, N7, N5, 'R1trimmed2break')\n",
    "        #file_R2 = create_filename(dir_sample, N7, N5, 'R2trimmed2break')\n",
    "\n",
    "    file_sam_genome_local = create_filename(dir_sample, N7, N5, 'break_trimmed_sam_genome_local')\n",
    "    file_sam_report_genome_local = create_filename(dir_sample, N7, N5, 'break_trimmed_sam_report_genome_local')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_sam_genome_local)):\n",
    "        os.mkdir(os.path.dirname(file_sam_genome_local))\n",
    "\n",
    "    file_bam_genome_local = create_filename(dir_sample, N7, N5, 'break_trimmed_bam_genome_local')\n",
    "    file_sorted_bam_genome_local = create_filename(dir_sample, N7, N5, 'break_trimmed_sorted_bam_genome_local')\n",
    "    # file_sorted_bai_genome_local = create_filename(dir_sample, N7, N5, 'sorted_bai_genome_local')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_bam_genome_local)):\n",
    "        os.mkdir(os.path.dirname(file_bam_genome_local))\n",
    "\n",
    "    # local alignment to the genome with bowtie2\n",
    "    initial_dir = os.getcwd()\n",
    "\n",
    "    bowtie2_command = ['bowtie2', '--local', '-p', str(ncpu),\n",
    "                       '-X', '5000', '-k', '2', '-x', assembly,\n",
    "                             '-U', file_R1, '-S', file_sam_genome_local]\n",
    "\n",
    "    handle_sam_report_genome_local = open(file_sam_report_genome_local, 'wb')\n",
    "\n",
    "    subprocess.call(bowtie2_command, stderr=handle_sam_report_genome_local)\n",
    "\n",
    "    handle_sam_report_genome_local.close()\n",
    "\n",
    "    # convert sam to bam\n",
    "    sam_to_bam_genome_local_command = ['samtools', 'view', '-Sb', file_sam_genome_local]\n",
    "\n",
    "    handle_file_bam_genome_local = open(file_bam_genome_local, 'wb')\n",
    "\n",
    "    subprocess.call(sam_to_bam_genome_local_command, stdout=handle_file_bam_genome_local)\n",
    "\n",
    "    # sort bam files\n",
    "    sort_bam_genome_local_command = ['samtools', 'sort', file_bam_genome_local, '-o', file_sorted_bam_genome_local]\n",
    "\n",
    "    subprocess.call(sort_bam_genome_local_command)\n",
    "\n",
    "    # Clean up\n",
    "    #os.remove(file_sam_genome_local)\n",
    "    os.remove(file_bam_genome_local)\n",
    "\n",
    "    # Create bam index files\n",
    "    create_bam_genome_local_index_command = ['samtools', 'index', file_sorted_bam_genome_local]\n",
    "    subprocess.call(create_bam_genome_local_index_command)\n",
    "\n",
    "    os.chdir(initial_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "align_afterbreak_genome_global(directory, amplicon_info, assembly)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "align_afterbreaks_genome_local(directory, 1, amplicon_info, assembly)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
