{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "import shutil\n",
    "\n",
    "from uditas.uditas_replace_helpers import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Pipeline1\n",
    "# This, for the most part, is the UDITAS pipeline modified to allow for REPLACE targeting. \n",
    "\n",
    "## There is a second pipeline below to do global alignments to quantify large deletions, translocatins, concatomers, etc by aligning the sequence past the break site to the entire genome+targeting sequence.\n",
    "\n",
    "## Inputs\n",
    "#### next few window are different inputs to make it run well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5\n"
     ]
    }
   ],
   "source": [
    "#things taken from original uditas args\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off\n",
    "\n",
    "############# Directory ############\n",
    "#this is a minimal directory of only ~300 files for quick debugging\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5/10000Reads'\n",
    "\n",
    "directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5'\n",
    "\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Master Number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this is for pipeline1 polb rev (row 1 in sheet \n",
      "\n",
      "primer_seq_plus_downstream ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                        Tn5_Pol50_pbR\n",
       "Sample                                                                               Polb\n",
       "description                                   MiniSeq_K562_PolbHIROS_50per_unsorted_polbR\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                      prE369\n",
       "I7_Index_ID                                                          P7_N702_SBS12nextera\n",
       "index_I1                                                                             N702\n",
       "barcode_I1                                                                       CGTACTAG\n",
       "I5_Index_ID                                                              P5_UMI_N50_Tn5-A\n",
       "index_I2                                                                             N501\n",
       "barcode_I2                                                                       GCGATCTA\n",
       "primer1_ForNest                                                                     prE87\n",
       "primer                                                                             prE364\n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                            42350386\n",
       "end                                                                              42350364\n",
       "strand                                                                                  -\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "genome_plus_targeting                                              hg38_plus_targetvector\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                            pE049_WO_redundant\n",
       "plasmid_sequence                        ttacataacttacggtaaatggcccgcctggctgaccgcccaacga...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 1, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number in list of \n",
    "\n",
    "#0 is polbRev\n",
    "#1 is polbRev \n",
    "#2 is mcherry_rev\n",
    "#3 is mcherry_fwd\n",
    "\n",
    "\n",
    "master_number = 1\n",
    "\n",
    "\n",
    "\n",
    "#######################################################################################\n",
    "#primer sequences for mispriming. Primer sequence + 12nt downstream \n",
    "#this will only be avialable for read2\n",
    "\n",
    "##### the sequences need to be capital for this to work########\n",
    "\n",
    "primerlist = [['TGGTTTTATTTCACCCCATGATAGTAATTATATCACT', 'this is for polb fwd (row 0 in sheet)'],\n",
    "              ['ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGC', 'this is for pipeline1 polb rev (row 1 in sheet'],\n",
    "              ['GTTAGCAGACTTCCTCTGCCCTCAATCTTTAAGAA', 'this is for pipeline1 mCherryRev, row2 in sheet'],\n",
    "              ['GCAGGAGCTCGTCGACCCATGGGGGCCCGCCCC', 'this is for pipeline1 mCherryfwd, row3 in sheet'],\n",
    "              ['TGGTTTTATTTCACCCCATGATAGTAATTATATCACT', 'this is for polb fwd WT (row 4 in sheet) was contaminated'],\n",
    "              ['ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGC', 'this is for pipeline1 WT polb rev (row 5 in sheet']]\n",
    "\n",
    "                \n",
    "\n",
    "primer_seq_plus_downstream = primerlist[master_number][0]          \n",
    "print(primerlist[master_number][1], '\\n')    \n",
    "print('primer_seq_plus_downstream', primer_seq_plus_downstream)\n",
    "\n",
    "\n",
    "#check what is printed below!!!\n",
    "\n",
    "#############################################################################################\n",
    "# Function to make the 'amplicon_info' list. Taking the line of our experiment csv file\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "\n",
    "#the amplicon info is related to the line on the csv file. It is indexed from 0. For PolbF we use 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "\n",
    "amplicon_info = get_csv_data(directory, master_number)\n",
    "\n",
    "###########check waht is printed below #######\n",
    "amplicon_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "assembly = amplicon_info['genome']\n",
    "file_genome_2bit = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly + '.2bit')\n",
    "print(file_genome_2bit)\n",
    "\n",
    "# BOWTIE2_INDEXES are needed for global alignments\n",
    "#not sure if this will work\n",
    "#normally in bash: export BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "\n",
    "#this is the refrence genome for Bowtie from Illumina. It does not have our targeting vector\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "\n",
    "#check the env variables\n",
    "#%env"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions to process the reads \n",
    "I imported them into the helper_replace python file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discard Mispriming Reads\n",
    "When you put a universal primer on the ends of everything, every mispriming event will amplify. An effect we normally don't deal with. I did nested PCR to reduce this. However 85% of the alignments in the UDITAS data I looked at seemed to be mispriming. They did all their blasting and analysis before removing mispriming. But to save computational power and remove error early on I will discard mispriming events. \n",
    "They discard these only for plasmid alignments analyze_alignments_plasmid for some reason which comes from the bam file.\n",
    "\n",
    "By eye it looks like 50-90% of my reads are correctly primed which is amaizng. Nesting helepd a lot.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reads_in_experiment_list_count 189785\n",
      " reads with good priming: 136172\n",
      "reads misprimed 53613\n"
     ]
    }
   ],
   "source": [
    "correct_priming(directory, amplicon_info, primer_seq_plus_downstream)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trim off illumina adapter from shorter reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Running the trimming\n",
    "\n",
    "trim_fastq(directory, amplicon_info, 0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running the reference plasmid creation\n",
    "create_plasmid_reference(directory, amplicon_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'replace'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Test the get_reaction_type is working\n",
    "#uditas_replace_helpers.get_reaction_type(amplicon_info)\n",
    "get_reaction_type(amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make the amplicons by calling the create_amplicon function\n",
    "\n",
    "create_amplicon(directory, amplicon_info, file_genome_2bit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try out the alignment to the plasmid\n",
    "align_plasmid_local(directory, amplicon_info, ncpu=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract the unmapped reads\n",
    "extract_unmapped_reads_plasmid(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### for some reason it throws errors when running from the python helper file so I just copied it here \n",
    "#analyze plasmid alignments\n",
    "\n",
    "def analyze_alignments_plasmid(dir_sample, amplicon_info, min_MAPQ, file_genome_2bit, do_plasmid):\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "        \n",
    "    exp_dir = create_filename(dir_sample, N7, N5, 'mainfolder')\n",
    "\n",
    "    file_UMI = create_filename(dir_sample, N7, N5, 'umifastqgz')\n",
    "    UMI_dict = create_barcode_dict(file_UMI)\n",
    "    \n",
    "    results_folder = os.path.join(exp_dir, 'results')\n",
    "    if not os.path.exists(results_folder):\n",
    "        os.mkdir(results_folder)\n",
    "\n",
    "    results_file = create_filename(dir_sample, N7, N5, 'results_plasmid')\n",
    "\n",
    "    if do_plasmid:\n",
    "        file_sorted_bam_plasmid_local = create_filename(dir_sample, N7, N5, 'sorted_bam_plasmid_local')\n",
    "\n",
    "        bam_in_alignment_file = pysam.AlignmentFile(file_sorted_bam_plasmid_local, 'rb')\n",
    "        bam_in = bam_in_alignment_file.fetch()\n",
    "\n",
    "        genome = twobitreader.TwoBitFile(file_genome_2bit)  # Load genome. Used for getting the sequences\n",
    "        \n",
    "        length_to_test = 15  # We check this number of bases after the primer\n",
    "        uditas_primer_length = amplicon_info['end'] - amplicon_info['start']\n",
    "        \n",
    "        if amplicon_info['strand'] == '+':  # This is the UDiTaS oligo strand\n",
    "            #I had to add int() command to make this work for some reason\n",
    "            seq_after_uditas_primer = genome[amplicon_info['chr']][int(amplicon_info['end']):int((amplicon_info['end'] + length_to_test))]\n",
    "            \n",
    "        elif amplicon_info['strand'] == '-':\n",
    "            seq_after_uditas_primer = reverse_complement(genome[amplicon_info['chr']][int((amplicon_info['start'] - length_to_test)):(int(amplicon_info['start']))])\n",
    "        n_max_mismatches = 2  # We allow this number of mismatches between the read and the sequence after the primer\n",
    "\n",
    "        names_list_plasmid_genome = []\n",
    "        UMI_list_plasmid_genome = []\n",
    "        names_list_plasmid_only = []\n",
    "        UMI_list_plasmid_only = []\n",
    "        \n",
    "        for read in bam_in:\n",
    "            if read.mapping_quality >= min_MAPQ and not read.is_unmapped and not read.is_secondary:\n",
    "                if read.is_read2:  # R2 is the UDiTaS primer\n",
    "                    if read.is_reverse:\n",
    "                        seq_test = reverse_complement(read.query_sequence)[int(uditas_primer_length):int((uditas_primer_length + length_to_test))]\n",
    "                    else:\n",
    "                        seq_test = read.query_sequence[int(uditas_primer_length): int(uditas_primer_length + length_to_test)]\n",
    "                    # Sometimes, after cutadapt we have a read shorter than uditas_primer_length + length_to_test\n",
    "                    # We skip those directly without calculating hamm_dist, which doesn't make sense\n",
    "                    if (len(seq_test) == len(seq_after_uditas_primer.upper()) and\n",
    "                        hamm_dist(seq_test, seq_after_uditas_primer.upper()) <= n_max_mismatches):\n",
    "                        # Reads for which the R2 has genomic sequence after the UDiTaS primer\n",
    "                        UMI_list_plasmid_genome.append(UMI_dict[read.query_name][0])\n",
    "                        names_list_plasmid_genome.append(read.query_name)\n",
    "                    else: # We put those short reads into the plasmid only bucket\n",
    "                        UMI_list_plasmid_only.append(UMI_dict[read.query_name][0])\n",
    "                        names_list_plasmid_only.append(read.query_name)\n",
    "\n",
    "        total_reads_plasmid_genome = len(set(names_list_plasmid_genome))\n",
    "        total_reads_collapsed_plasmid_genome = len(set(UMI_list_plasmid_genome))\n",
    "        total_reads_plasmid_only = len(set(names_list_plasmid_only))\n",
    "        total_reads_collapsed_plasmid_only = len(set(UMI_list_plasmid_only))\n",
    "\n",
    "        results_df = pd.DataFrame({'target_plus_plasmid_total_reads': [total_reads_plasmid_genome],\n",
    "                                   'target_plus_plasmid_total_reads_collapsed': [total_reads_collapsed_plasmid_genome],\n",
    "                                   'plasmid_only_total_reads': [total_reads_plasmid_only],\n",
    "                                   'plasmid_only_total_reads_collapsed': [total_reads_collapsed_plasmid_only]\n",
    "                                   },\n",
    "                                  columns=['target_plus_plasmid_total_reads',\n",
    "                                           'target_plus_plasmid_total_reads_collapsed',\n",
    "                                           'plasmid_only_total_reads',\n",
    "                                           'plasmid_only_total_reads_collapsed'])\n",
    "    else:\n",
    "        results_df = pd.DataFrame(index=np.arange(1),\n",
    "                                  columns=['target_plus_plasmid_total_reads',\n",
    "                                           'target_plus_plasmid_total_reads_collapsed',\n",
    "                                           'plasmid_only_total_reads',\n",
    "                                           'plasmid_only_total_reads_collapsed'])\n",
    "\n",
    "    results_df.to_excel(results_file)\n",
    "\n",
    "    return results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3095</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target_plus_plasmid_total_reads  target_plus_plasmid_total_reads_collapsed  \\\n",
       "0                                0                                          0   \n",
       "\n",
       "   plasmid_only_total_reads  plasmid_only_total_reads_collapsed  \n",
       "0                      3095                                  67  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#run the plasmid analysis to count plasmid integration events\n",
    "result_plasmid_df = analyze_alignments_plasmid(directory, amplicon_info, min_MAPQ, file_genome_2bit, True)\n",
    "result_plasmid_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#align against our suite of amplicons\n",
    "# took about 5 min with 300k reads\n",
    "align_amplicon(directory, amplicon_info, check_plasmid_insertions, ncpu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this will extract unmap reads new folder (files that did not align to the predicted structural variants)\n",
    "extract_unmapped_reads_amplicons(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This section is to do the analysis of the amplicon alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reads before mispriming cleanup 407906\n",
      "number of unique umis before mispriming cleanup 174757\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>95912</td>\n",
       "      <td>2160</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   all_amplicons_total_reads  all_amplicons_total_reads_collapsed\n",
       "0                      95912                                 2160"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_reads_in_all_amplicons_df = analyze_alignments_all_amplicons(directory, amplicon_info, min_MAPQ, min_AS)\n",
    "result_reads_in_all_amplicons_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_cut1_total_reads</th>\n",
       "      <th>wt_cut1_total_indels</th>\n",
       "      <th>wt_cut1_total_deletions</th>\n",
       "      <th>wt_cut1_total_insertions</th>\n",
       "      <th>wt_cut1_total_reads_collapsed</th>\n",
       "      <th>wt_cut1_total_indels_collapsed</th>\n",
       "      <th>wt_cut1_total_deletions_collapsed</th>\n",
       "      <th>wt_cut1_total_insertions_collapsed</th>\n",
       "      <th>wt_cut2_total_reads</th>\n",
       "      <th>wt_cut2_total_indels</th>\n",
       "      <th>...</th>\n",
       "      <th>1a_1a_cut1_total_insertions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_reads</th>\n",
       "      <th>2b_2b_cut1_total_indels</th>\n",
       "      <th>2b_2b_cut1_total_deletions</th>\n",
       "      <th>2b_2b_cut1_total_insertions</th>\n",
       "      <th>2b_2b_cut1_total_reads_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_indels_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_deletions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_insertions_collapsed</th>\n",
       "      <th>median_fragment_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>165.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 145 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   wt_cut1_total_reads  wt_cut1_total_indels  wt_cut1_total_deletions  \\\n",
       "0                    0                     0                        0   \n",
       "\n",
       "   wt_cut1_total_insertions  wt_cut1_total_reads_collapsed  \\\n",
       "0                         0                              0   \n",
       "\n",
       "   wt_cut1_total_indels_collapsed  wt_cut1_total_deletions_collapsed  \\\n",
       "0                               0                                  0   \n",
       "\n",
       "   wt_cut1_total_insertions_collapsed  wt_cut2_total_reads  \\\n",
       "0                                   0                    0   \n",
       "\n",
       "   wt_cut2_total_indels  ...  1a_1a_cut1_total_insertions_collapsed  \\\n",
       "0                     0  ...                                      0   \n",
       "\n",
       "   2b_2b_cut1_total_reads  2b_2b_cut1_total_indels  \\\n",
       "0                       0                        0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions  2b_2b_cut1_total_insertions  \\\n",
       "0                           0                            0   \n",
       "\n",
       "   2b_2b_cut1_total_reads_collapsed  2b_2b_cut1_total_indels_collapsed  \\\n",
       "0                                 0                                  0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions_collapsed  \\\n",
       "0                                     0   \n",
       "\n",
       "   2b_2b_cut1_total_insertions_collapsed  median_fragment_size  \n",
       "0                                      0                 165.0  \n",
       "\n",
       "[1 rows x 145 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#took \n",
    "result_amplicon_df = analyze_alignments(directory, amplicon_info, window_size, amplicon_window_around_cut, min_MAPQ, min_AS)\n",
    "result_amplicon_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is the Global alignment of the remaining reads\n",
    "\n",
    "#### What I need here is to modify it so that "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run the allignment\n",
    "assembly = amplicon_info['genome']\n",
    "align_genome_global(directory, amplicon_info, assembly)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of alignments that were attempted to align 34949\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16490</td>\n",
       "      <td>414</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genomewide_total_reads  genomewide_total_reads_collapsed  \\\n",
       "0                   16490                               414   \n",
       "\n",
       "   genomewide_target_only_reads  genomewide_target_only_reads_collapsed  \n",
       "0                             0                                       0  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Do the global alignment analysis\n",
    "results_genome_global_df = analyze_alignments_genome_global(directory, amplicon_info, min_MAPQ, min_AS,  file_genome_2bit)\n",
    "results_genome_global_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summerize data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_count = pd.DataFrame()\n",
    "read_count.loc[0,'read_count'] = count_reads(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_count</th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "      <th>total_aligned</th>\n",
       "      <th>total_aligned_collapsed</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>percent_aligned_all_amplicons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>222201.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3095</td>\n",
       "      <td>67</td>\n",
       "      <td>87042</td>\n",
       "      <td>2020</td>\n",
       "      <td>16490</td>\n",
       "      <td>414</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>106627</td>\n",
       "      <td>2501</td>\n",
       "      <td>47.986733</td>\n",
       "      <td>81.632232</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   read_count  target_plus_plasmid_total_reads  \\\n",
       "0    222201.0                                0   \n",
       "\n",
       "   target_plus_plasmid_total_reads_collapsed  plasmid_only_total_reads  \\\n",
       "0                                          0                      3095   \n",
       "\n",
       "   plasmid_only_total_reads_collapsed  all_amplicons_total_reads  \\\n",
       "0                                  67                      87042   \n",
       "\n",
       "   all_amplicons_total_reads_collapsed  genomewide_total_reads  \\\n",
       "0                                 2020                   16490   \n",
       "\n",
       "   genomewide_total_reads_collapsed  genomewide_target_only_reads  \\\n",
       "0                               414                             0   \n",
       "\n",
       "   genomewide_target_only_reads_collapsed  total_aligned  \\\n",
       "0                                       0         106627   \n",
       "\n",
       "   total_aligned_collapsed  percent_aligned  percent_aligned_all_amplicons  \n",
       "0                     2501        47.986733                      81.632232  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#summerize the data\n",
    "\n",
    "summary_all_alignments = get_summary_all_alignments(directory, amplicon_info, read_count.loc[[0]], result_plasmid_df, result_reads_in_all_amplicons_df, results_genome_global_df)\n",
    "summary_all_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NGS_req-ID</th>\n",
       "      <th>name</th>\n",
       "      <th>Sample</th>\n",
       "      <th>description</th>\n",
       "      <th>Control sample (Y/N)</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Dilution</th>\n",
       "      <th>Cell name_type</th>\n",
       "      <th>UMI_Len</th>\n",
       "      <th>IndexI7Primer</th>\n",
       "      <th>...</th>\n",
       "      <th>doner_tail_tail_cut1_total_reads_percent</th>\n",
       "      <th>doner_head_tail_cut1_total_reads_percent</th>\n",
       "      <th>doner_head_head_cut1_total_reads_percent</th>\n",
       "      <th>1a_1a_cut1_total_reads_percent</th>\n",
       "      <th>2b_2b_cut1_total_reads_percent</th>\n",
       "      <th>target_plus_plasmid_total_reads_percent</th>\n",
       "      <th>plasmid_only_total_reads_percent</th>\n",
       "      <th>total_aligned_junctions_collapsed</th>\n",
       "      <th>Type</th>\n",
       "      <th>Percent Editing</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_collapsed_per...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbF</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE368</td>\n",
       "      <td>...</td>\n",
       "      <td>1.033386</td>\n",
       "      <td>38.022922</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.672021</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>3.390688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_polbR</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE369</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_mCh</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_Pol50_bpa</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE371</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbF</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE372</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>Tn5_WT_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>prE373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_collapsed_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 224 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     NGS_req-ID           name Sample  \\\n",
       "0    A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "1    A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "2    A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "3    A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "4    A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "5    A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "6    A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "7    A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "8    A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "9    A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "10   A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "11   A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "12   A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "13   A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "14   A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "15   A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "16   A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "17   A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "18   A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "19   A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "20   A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "21   A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "22   A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "23   A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "24   A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "25   A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "26   A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "27   A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "28   A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "29   A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "..          ...            ...    ...   \n",
       "90   A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "91   A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "92   A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "93   A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "94   A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "95   A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "96   A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "97   A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "98   A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "99   A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "100  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "101  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "102  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "103  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "104  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "105  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "106  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "107  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "108  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "109  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "110  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "111  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "112  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "113  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "114  A000H2GWGF  Tn5_Pol50_pbF   Polb   \n",
       "115  A000H2GWGF  Tn5_Pol50_pbR   Polb   \n",
       "116  A000H2GWGF  Tn5_Pol50_mCh   Polb   \n",
       "117  A000H2GWGF  Tn5_Pol50_bpa   Polb   \n",
       "118  A000H2GWGF     Tn5_WT_pbF   Polb   \n",
       "119  A000H2GWGF     Tn5_WT_pbR   Polb   \n",
       "\n",
       "                                           description Control sample (Y/N)  \\\n",
       "0          MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "1          MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "2    MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "3    MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "4    MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "5    MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "6          MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "7          MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "8    MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "9    MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "10   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "11   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "12         MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "13         MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "14   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "15   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "16   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "17   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "18         MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "19         MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "20   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "21   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "22   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "23   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "24         MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "25         MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "26   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "27   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "28   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "29   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "..                                                 ...                  ...   \n",
       "90         MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "91         MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "92   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "93   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "94   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "95   MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "96         MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "97         MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "98   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "99   MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "100  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "101  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "102        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "103        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "104  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "105  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "106  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "107  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "108        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "109        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "110  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "111  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "112  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "113  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "114        MiniSeq_K562_PolbHIROS_50per_unsorted_polbF                    N   \n",
       "115        MiniSeq_K562_PolbHIROS_50per_unsorted_polbR                    N   \n",
       "116  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherryp...                    N   \n",
       "117  MiniSeq_K562_PolbHIROS_50per_unsorted_mCherry_Fwd                    N   \n",
       "118  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "119  MiniSeq_K562_PolbHIROS_Wtcontaminated_w_50per_...                    N   \n",
       "\n",
       "     Notes  Dilution Cell name_type     UMI_Len IndexI7Primer  ...  \\\n",
       "0      NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "1      NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "2      NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "3      NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "4      NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "5      NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "6      NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "7      NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "8      NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "9      NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "10     NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "11     NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "12     NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "13     NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "14     NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "15     NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "16     NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "17     NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "18     NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "19     NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "20     NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "21     NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "22     NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "23     NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "24     NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "25     NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "26     NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "27     NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "28     NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "29     NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "..     ...       ...            ...         ...           ...  ...   \n",
       "90     NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "91     NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "92     NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "93     NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "94     NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "95     NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "96     NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "97     NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "98     NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "99     NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "100    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "101    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "102    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "103    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "104    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "105    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "106    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "107    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "108    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "109    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "110    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "111    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "112    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "113    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "114    NaN       NaN           K562  NNNNNNNNNN        prE368  ...   \n",
       "115    NaN       NaN           K562  NNNNNNNNNN        prE369  ...   \n",
       "116    NaN       NaN           K562  NNNNNNNNNN        prE370  ...   \n",
       "117    NaN       NaN           K562  NNNNNNNNNN        prE371  ...   \n",
       "118    NaN       NaN           K562  NNNNNNNNNN        prE372  ...   \n",
       "119    NaN       NaN           K562  NNNNNNNNNN        prE373  ...   \n",
       "\n",
       "    doner_tail_tail_cut1_total_reads_percent  \\\n",
       "0                                   1.033386   \n",
       "1                                        NaN   \n",
       "2                                        NaN   \n",
       "3                                        NaN   \n",
       "4                                        NaN   \n",
       "5                                        NaN   \n",
       "6                                   1.033386   \n",
       "7                                        NaN   \n",
       "8                                        NaN   \n",
       "9                                        NaN   \n",
       "10                                       NaN   \n",
       "11                                       NaN   \n",
       "12                                  1.033386   \n",
       "13                                       NaN   \n",
       "14                                       NaN   \n",
       "15                                       NaN   \n",
       "16                                       NaN   \n",
       "17                                       NaN   \n",
       "18                                  1.033386   \n",
       "19                                       NaN   \n",
       "20                                       NaN   \n",
       "21                                       NaN   \n",
       "22                                       NaN   \n",
       "23                                       NaN   \n",
       "24                                  1.033386   \n",
       "25                                       NaN   \n",
       "26                                       NaN   \n",
       "27                                       NaN   \n",
       "28                                       NaN   \n",
       "29                                       NaN   \n",
       "..                                       ...   \n",
       "90                                  1.033386   \n",
       "91                                       NaN   \n",
       "92                                       NaN   \n",
       "93                                       NaN   \n",
       "94                                       NaN   \n",
       "95                                       NaN   \n",
       "96                                  1.033386   \n",
       "97                                       NaN   \n",
       "98                                       NaN   \n",
       "99                                       NaN   \n",
       "100                                      NaN   \n",
       "101                                      NaN   \n",
       "102                                 1.033386   \n",
       "103                                      NaN   \n",
       "104                                      NaN   \n",
       "105                                      NaN   \n",
       "106                                      NaN   \n",
       "107                                      NaN   \n",
       "108                                 1.033386   \n",
       "109                                      NaN   \n",
       "110                                      NaN   \n",
       "111                                      NaN   \n",
       "112                                      NaN   \n",
       "113                                      NaN   \n",
       "114                                 1.033386   \n",
       "115                                      NaN   \n",
       "116                                      NaN   \n",
       "117                                      NaN   \n",
       "118                                      NaN   \n",
       "119                                      NaN   \n",
       "\n",
       "    doner_head_tail_cut1_total_reads_percent  \\\n",
       "0                                  38.022922   \n",
       "1                                        NaN   \n",
       "2                                        NaN   \n",
       "3                                        NaN   \n",
       "4                                        NaN   \n",
       "5                                        NaN   \n",
       "6                                  38.022922   \n",
       "7                                        NaN   \n",
       "8                                        NaN   \n",
       "9                                        NaN   \n",
       "10                                       NaN   \n",
       "11                                       NaN   \n",
       "12                                 38.022922   \n",
       "13                                       NaN   \n",
       "14                                       NaN   \n",
       "15                                       NaN   \n",
       "16                                       NaN   \n",
       "17                                       NaN   \n",
       "18                                 38.022922   \n",
       "19                                       NaN   \n",
       "20                                       NaN   \n",
       "21                                       NaN   \n",
       "22                                       NaN   \n",
       "23                                       NaN   \n",
       "24                                 38.022922   \n",
       "25                                       NaN   \n",
       "26                                       NaN   \n",
       "27                                       NaN   \n",
       "28                                       NaN   \n",
       "29                                       NaN   \n",
       "..                                       ...   \n",
       "90                                 38.022922   \n",
       "91                                       NaN   \n",
       "92                                       NaN   \n",
       "93                                       NaN   \n",
       "94                                       NaN   \n",
       "95                                       NaN   \n",
       "96                                 38.022922   \n",
       "97                                       NaN   \n",
       "98                                       NaN   \n",
       "99                                       NaN   \n",
       "100                                      NaN   \n",
       "101                                      NaN   \n",
       "102                                38.022922   \n",
       "103                                      NaN   \n",
       "104                                      NaN   \n",
       "105                                      NaN   \n",
       "106                                      NaN   \n",
       "107                                      NaN   \n",
       "108                                38.022922   \n",
       "109                                      NaN   \n",
       "110                                      NaN   \n",
       "111                                      NaN   \n",
       "112                                      NaN   \n",
       "113                                      NaN   \n",
       "114                                38.022922   \n",
       "115                                      NaN   \n",
       "116                                      NaN   \n",
       "117                                      NaN   \n",
       "118                                      NaN   \n",
       "119                                      NaN   \n",
       "\n",
       "    doner_head_head_cut1_total_reads_percent 1a_1a_cut1_total_reads_percent  \\\n",
       "0                                        0.0                            0.0   \n",
       "1                                        NaN                            NaN   \n",
       "2                                        NaN                            NaN   \n",
       "3                                        NaN                            NaN   \n",
       "4                                        NaN                            NaN   \n",
       "5                                        NaN                            NaN   \n",
       "6                                        0.0                            0.0   \n",
       "7                                        NaN                            NaN   \n",
       "8                                        NaN                            NaN   \n",
       "9                                        NaN                            NaN   \n",
       "10                                       NaN                            NaN   \n",
       "11                                       NaN                            NaN   \n",
       "12                                       0.0                            0.0   \n",
       "13                                       NaN                            NaN   \n",
       "14                                       NaN                            NaN   \n",
       "15                                       NaN                            NaN   \n",
       "16                                       NaN                            NaN   \n",
       "17                                       NaN                            NaN   \n",
       "18                                       0.0                            0.0   \n",
       "19                                       NaN                            NaN   \n",
       "20                                       NaN                            NaN   \n",
       "21                                       NaN                            NaN   \n",
       "22                                       NaN                            NaN   \n",
       "23                                       NaN                            NaN   \n",
       "24                                       0.0                            0.0   \n",
       "25                                       NaN                            NaN   \n",
       "26                                       NaN                            NaN   \n",
       "27                                       NaN                            NaN   \n",
       "28                                       NaN                            NaN   \n",
       "29                                       NaN                            NaN   \n",
       "..                                       ...                            ...   \n",
       "90                                       0.0                            0.0   \n",
       "91                                       NaN                            NaN   \n",
       "92                                       NaN                            NaN   \n",
       "93                                       NaN                            NaN   \n",
       "94                                       NaN                            NaN   \n",
       "95                                       NaN                            NaN   \n",
       "96                                       0.0                            0.0   \n",
       "97                                       NaN                            NaN   \n",
       "98                                       NaN                            NaN   \n",
       "99                                       NaN                            NaN   \n",
       "100                                      NaN                            NaN   \n",
       "101                                      NaN                            NaN   \n",
       "102                                      0.0                            0.0   \n",
       "103                                      NaN                            NaN   \n",
       "104                                      NaN                            NaN   \n",
       "105                                      NaN                            NaN   \n",
       "106                                      NaN                            NaN   \n",
       "107                                      NaN                            NaN   \n",
       "108                                      0.0                            0.0   \n",
       "109                                      NaN                            NaN   \n",
       "110                                      NaN                            NaN   \n",
       "111                                      NaN                            NaN   \n",
       "112                                      NaN                            NaN   \n",
       "113                                      NaN                            NaN   \n",
       "114                                      0.0                            0.0   \n",
       "115                                      NaN                            NaN   \n",
       "116                                      NaN                            NaN   \n",
       "117                                      NaN                            NaN   \n",
       "118                                      NaN                            NaN   \n",
       "119                                      NaN                            NaN   \n",
       "\n",
       "    2b_2b_cut1_total_reads_percent target_plus_plasmid_total_reads_percent  \\\n",
       "0                              0.0                                     0.0   \n",
       "1                              NaN                                     NaN   \n",
       "2                              NaN                                     NaN   \n",
       "3                              NaN                                     NaN   \n",
       "4                              NaN                                     NaN   \n",
       "5                              NaN                                     NaN   \n",
       "6                              0.0                                     0.0   \n",
       "7                              NaN                                     NaN   \n",
       "8                              NaN                                     NaN   \n",
       "9                              NaN                                     NaN   \n",
       "10                             NaN                                     NaN   \n",
       "11                             NaN                                     NaN   \n",
       "12                             0.0                                     0.0   \n",
       "13                             NaN                                     NaN   \n",
       "14                             NaN                                     NaN   \n",
       "15                             NaN                                     NaN   \n",
       "16                             NaN                                     NaN   \n",
       "17                             NaN                                     NaN   \n",
       "18                             0.0                                     0.0   \n",
       "19                             NaN                                     NaN   \n",
       "20                             NaN                                     NaN   \n",
       "21                             NaN                                     NaN   \n",
       "22                             NaN                                     NaN   \n",
       "23                             NaN                                     NaN   \n",
       "24                             0.0                                     0.0   \n",
       "25                             NaN                                     NaN   \n",
       "26                             NaN                                     NaN   \n",
       "27                             NaN                                     NaN   \n",
       "28                             NaN                                     NaN   \n",
       "29                             NaN                                     NaN   \n",
       "..                             ...                                     ...   \n",
       "90                             0.0                                     0.0   \n",
       "91                             NaN                                     NaN   \n",
       "92                             NaN                                     NaN   \n",
       "93                             NaN                                     NaN   \n",
       "94                             NaN                                     NaN   \n",
       "95                             NaN                                     NaN   \n",
       "96                             0.0                                     0.0   \n",
       "97                             NaN                                     NaN   \n",
       "98                             NaN                                     NaN   \n",
       "99                             NaN                                     NaN   \n",
       "100                            NaN                                     NaN   \n",
       "101                            NaN                                     NaN   \n",
       "102                            0.0                                     0.0   \n",
       "103                            NaN                                     NaN   \n",
       "104                            NaN                                     NaN   \n",
       "105                            NaN                                     NaN   \n",
       "106                            NaN                                     NaN   \n",
       "107                            NaN                                     NaN   \n",
       "108                            0.0                                     0.0   \n",
       "109                            NaN                                     NaN   \n",
       "110                            NaN                                     NaN   \n",
       "111                            NaN                                     NaN   \n",
       "112                            NaN                                     NaN   \n",
       "113                            NaN                                     NaN   \n",
       "114                            0.0                                     0.0   \n",
       "115                            NaN                                     NaN   \n",
       "116                            NaN                                     NaN   \n",
       "117                            NaN                                     NaN   \n",
       "118                            NaN                                     NaN   \n",
       "119                            NaN                                     NaN   \n",
       "\n",
       "    plasmid_only_total_reads_percent total_aligned_junctions_collapsed  \\\n",
       "0                           3.672021                            1976.0   \n",
       "1                                NaN                               NaN   \n",
       "2                                NaN                               NaN   \n",
       "3                                NaN                               NaN   \n",
       "4                                NaN                               NaN   \n",
       "5                                NaN                               NaN   \n",
       "6                           3.672021                            1976.0   \n",
       "7                                NaN                               NaN   \n",
       "8                                NaN                               NaN   \n",
       "9                                NaN                               NaN   \n",
       "10                               NaN                               NaN   \n",
       "11                               NaN                               NaN   \n",
       "12                          3.672021                            1976.0   \n",
       "13                               NaN                               NaN   \n",
       "14                               NaN                               NaN   \n",
       "15                               NaN                               NaN   \n",
       "16                               NaN                               NaN   \n",
       "17                               NaN                               NaN   \n",
       "18                          3.672021                            1976.0   \n",
       "19                               NaN                               NaN   \n",
       "20                               NaN                               NaN   \n",
       "21                               NaN                               NaN   \n",
       "22                               NaN                               NaN   \n",
       "23                               NaN                               NaN   \n",
       "24                          3.672021                            1976.0   \n",
       "25                               NaN                               NaN   \n",
       "26                               NaN                               NaN   \n",
       "27                               NaN                               NaN   \n",
       "28                               NaN                               NaN   \n",
       "29                               NaN                               NaN   \n",
       "..                               ...                               ...   \n",
       "90                          3.672021                            1976.0   \n",
       "91                               NaN                               NaN   \n",
       "92                               NaN                               NaN   \n",
       "93                               NaN                               NaN   \n",
       "94                               NaN                               NaN   \n",
       "95                               NaN                               NaN   \n",
       "96                          3.672021                            1976.0   \n",
       "97                               NaN                               NaN   \n",
       "98                               NaN                               NaN   \n",
       "99                               NaN                               NaN   \n",
       "100                              NaN                               NaN   \n",
       "101                              NaN                               NaN   \n",
       "102                         3.672021                            1976.0   \n",
       "103                              NaN                               NaN   \n",
       "104                              NaN                               NaN   \n",
       "105                              NaN                               NaN   \n",
       "106                              NaN                               NaN   \n",
       "107                              NaN                               NaN   \n",
       "108                         3.672021                            1976.0   \n",
       "109                              NaN                               NaN   \n",
       "110                              NaN                               NaN   \n",
       "111                              NaN                               NaN   \n",
       "112                              NaN                               NaN   \n",
       "113                              NaN                               NaN   \n",
       "114                         3.672021                            1976.0   \n",
       "115                              NaN                               NaN   \n",
       "116                              NaN                               NaN   \n",
       "117                              NaN                               NaN   \n",
       "118                              NaN                               NaN   \n",
       "119                              NaN                               NaN   \n",
       "\n",
       "                                                  Type Percent Editing  \n",
       "0                wt_cut1_total_reads_collapsed_percent        0.000000  \n",
       "1                wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "2                wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "3                wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "4                wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "5                wt_cut1_total_reads_collapsed_percent             NaN  \n",
       "6                wt_cut2_total_reads_collapsed_percent        0.000000  \n",
       "7                wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "8                wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "9                wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "10               wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "11               wt_cut2_total_reads_collapsed_percent             NaN  \n",
       "12   large_deletion_cut1_total_reads_collapsed_percent        0.000000  \n",
       "13   large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "14   large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "15   large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "16   large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "17   large_deletion_cut1_total_reads_collapsed_percent             NaN  \n",
       "18   large_inversion_cut1_total_reads_collapsed_per...        0.000000  \n",
       "19   large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "20   large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "21   large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "22   large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "23   large_inversion_cut1_total_reads_collapsed_per...             NaN  \n",
       "24   large_inversion_cut2_total_reads_collapsed_per...        0.000000  \n",
       "25   large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "26   large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "27   large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "28   large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "29   large_inversion_cut2_total_reads_collapsed_per...             NaN  \n",
       "..                                                 ...             ...  \n",
       "90   doner_head_head_cut1_total_reads_collapsed_per...        0.000000  \n",
       "91   doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "92   doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "93   doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "94   doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "95   doner_head_head_cut1_total_reads_collapsed_per...             NaN  \n",
       "96            1a_1a_cut1_total_reads_collapsed_percent        0.000000  \n",
       "97            1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "98            1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "99            1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "100           1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "101           1a_1a_cut1_total_reads_collapsed_percent             NaN  \n",
       "102           2b_2b_cut1_total_reads_collapsed_percent        0.000000  \n",
       "103           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "104           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "105           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "106           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "107           2b_2b_cut1_total_reads_collapsed_percent             NaN  \n",
       "108  target_plus_plasmid_total_reads_collapsed_percent        0.000000  \n",
       "109  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "110  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "111  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "112  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "113  target_plus_plasmid_total_reads_collapsed_percent             NaN  \n",
       "114         plasmid_only_total_reads_collapsed_percent        3.390688  \n",
       "115         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "116         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "117         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "118         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "119         plasmid_only_total_reads_collapsed_percent             NaN  \n",
       "\n",
       "[120 rows x 224 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##### you can't run this cell alone. It requires analysis higher in the pipeline\n",
    "\n",
    "results_alignments_junction = pd.concat([result_amplicon_df, result_plasmid_df], axis=1)\n",
    "\n",
    "sample_info_filename = os.path.join(directory, 'sample_info.csv')\n",
    "\n",
    "experiments = pd.read_csv(sample_info_filename)\n",
    "\n",
    "N7 = amplicon_info['index_I1']\n",
    "N5 = amplicon_info['index_I2']\n",
    "\n",
    "results_summary = summarize_results(results_alignments_junction)\n",
    "results_summary\n",
    "#this outputs it in a directory above the N501_701 folder as a summary file. Need to label it or it will be overwritten by next tone\n",
    "results_summary_with_experiments = pd.concat([experiments, results_summary], axis=1)\n",
    "results_summary_with_experiments.to_excel(os.path.join(directory, N7 + '_' + N5, N7 + '_' + N5 + '_' + 'results_summary.xlsx'))\n",
    "\n",
    "results_pivot = melt_results(results_summary_with_experiments)\n",
    "results_pivot.to_excel(os.path.join(directory, N7 + '_' + N5, N7 + '_' + N5+'_'+'results_summary_pivot.xlsx'))\n",
    "results_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline2\n",
    "#  This is a second pipeline. Different than Uditas but built in a simliar form. \n",
    "## It uses the read distal to the gene specic primer. This read covers the region downstream of the break. After clipping off anything remaing upstream of the break we make a global alignment to look at genomic integrations and translocations. This also quantifies the alignments.\n",
    "\n",
    "\n",
    "### 0. Use correct_prime function in pipeline1 to arrive at reads that were correctly primed.\n",
    "### 1. Trim correctly primed reads to the break site  \n",
    "### 2. Only use a single end of the pair end reads (distal gene specific primer)\n",
    "### 3. Use bowtie2 genome that has the targeting seq in it added\n",
    "### 4. Do global end-to-end alignment\n",
    "### 5. Do a local alignment\n",
    "### 6. process bam file and pull out the high mapQ reads\n",
    "\n",
    "The idea here is to trim the sequence up to the cut on both reads and then do a global alignment end-to-end agasint the human genome.\n",
    "\n",
    "The idea here is to take the files that are 'trimmed' and 'correct priming' checking. Then to do another round of trimming to get down to the cut site. I need to make another folder for 'trimmed_to_cut'\n",
    "\n",
    "For Tn5 setup the 'Read 1' contains the sequence away from the gene specific primer. So it is the read we should concentration on if we don't do paired end.\n",
    "\n",
    "For LAM the 'Read 2' is the the sequence distal to the gene specific primer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Need to make an new bowtie2 index file that includes targeting for alignment agains the whole genome so it is all in one sheet together. \n",
    "\n",
    "### Other option would be to align it to the amplicon, extract unaligned files and then align to the genome but seems cleaner this way. \n",
    "#### Ideally every sequence is unique between the targeting vector and genome.\n",
    "\n",
    "\n",
    "1. Build your fastas of interest and label .fa files.\n",
    "    1. You need fasta of hg38 or reference genome. You can pull this from downloaded bowtie indexed sampels and then use the following command to turn the index into a fasta file: bowtie2-inspect hg38 > hg38.fa   \n",
    "    2. Put all the fasta files in the same folder\n",
    "2. index the files with bowtie\n",
    "    1. use the command bowtie2-build -f pE049,pe038_mc.fa,hg38.fa -p hg38_plus_targetvectorandplasmid\n",
    "    2. this has the -p to make it take less ram in my case.\n",
    "    3. In this case it adds the hg38 and the minicircle targeting file together\n",
    "    4. I have a Intel® Core™ i7-5500U CPU @ 2.40GHz × 4 with 15.1 GiB ram and it required about 13.8 gigs of ram and 2 hours to do a hg38+small fasta index\n",
    "    5. be sure to pay attention to the name of the new indexed file. \"hg38_plus_targetvector\" in example above. Add it to the sample_info.csv sheet. Under the tab 'genome_plus_targeting'.\n",
    "    6. you can check it indexed correcctly: bowtie2-inspect -s hg38_plus_targetvector\n",
    "    \n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A simple loop for runnign pipeline 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38_plus_targetvector.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
      "the directory is: /media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5\n",
      "running row 0\n",
      "this is for pipeline2 polb fwd (row 0 in sheet) \n",
      "\n",
      "minimum length after trimming is: 35\n",
      "tn5 analysis\n",
      "tn5 analysis\n",
      "sam file deleted\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N701 _ N501\n",
      "doing analysis on the global alignment N701 _ N501\n",
      "   %_target_site_reads  %_target_vector_reads  %_translocations  \\\n",
      "0            70.912312              28.807988            0.2797   \n",
      "\n",
      "   %_target_site_reads_collapsed  %_target_vector_reads_collapsed  \\\n",
      "0                      71.733871                        28.104839   \n",
      "\n",
      "   %_translocations_collapsed  alignments_target_site_all  \\\n",
      "0                     0.16129                      127272   \n",
      "\n",
      "   alignments_target_vector_all  alignments_translocations_all  \\\n",
      "0                         51704                            502   \n",
      "\n",
      "   collapsed_alignments_target_site  collapsed_alignments_target_vector  \\\n",
      "0                              1779                                 697   \n",
      "\n",
      "   collapsed_alignments_translocations  total aligned reads filtered  \\\n",
      "0                                    4                      179478.0   \n",
      "\n",
      "   total collpased reads filtered  all_reads_unfiltered  \\\n",
      "0                          2480.0                183138   \n",
      "\n",
      "   all_reads_unfiltered_collapsed  \n",
      "0                            2458  \n",
      "running row 1\n",
      "this is for pipeline2 polb rev (row 1 in sheet) \n",
      "\n",
      "minimum length after trimming is: 35\n",
      "tn5 analysis\n",
      "tn5 analysis\n",
      "sam file deleted\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N702 _ N501\n",
      "doing analysis on the global alignment N702 _ N501\n",
      "   %_target_site_reads  %_target_vector_reads  %_translocations  \\\n",
      "0            45.125983              54.670573          0.203444   \n",
      "\n",
      "   %_target_site_reads_collapsed  %_target_vector_reads_collapsed  \\\n",
      "0                      47.689625                        46.294682   \n",
      "\n",
      "   %_translocations_collapsed  alignments_target_site_all  \\\n",
      "0                    6.015693                       44362   \n",
      "\n",
      "   alignments_target_vector_all  alignments_translocations_all  \\\n",
      "0                         53745                            200   \n",
      "\n",
      "   collapsed_alignments_target_site  collapsed_alignments_target_vector  \\\n",
      "0                              1094                                1062   \n",
      "\n",
      "   collapsed_alignments_translocations  total aligned reads filtered  \\\n",
      "0                                  138                       98307.0   \n",
      "\n",
      "   total collpased reads filtered  all_reads_unfiltered  \\\n",
      "0                          2294.0                100106   \n",
      "\n",
      "   all_reads_unfiltered_collapsed  \n",
      "0                            2289  \n",
      "running row 2\n",
      "this is for pipeline2 mCherryRev, row2 in sheet \n",
      "\n",
      "minimum length after trimming is: 35\n",
      "tn5 analysis\n",
      "tn5 analysis\n",
      "sam file deleted\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N703 _ N501\n",
      "doing analysis on the global alignment N703 _ N501\n",
      "   %_target_site_reads  %_target_vector_reads  %_translocations  \\\n",
      "0            25.233041              57.430649          17.33631   \n",
      "\n",
      "   %_target_site_reads_collapsed  %_target_vector_reads_collapsed  \\\n",
      "0                      24.637681                        57.821089   \n",
      "\n",
      "   %_translocations_collapsed  alignments_target_site_all  \\\n",
      "0                   17.541229                       17974   \n",
      "\n",
      "   alignments_target_vector_all  alignments_translocations_all  \\\n",
      "0                         40909                          12349   \n",
      "\n",
      "   collapsed_alignments_target_site  collapsed_alignments_target_vector  \\\n",
      "0                               493                                1157   \n",
      "\n",
      "   collapsed_alignments_translocations  total aligned reads filtered  \\\n",
      "0                                  351                       71232.0   \n",
      "\n",
      "   total collpased reads filtered  all_reads_unfiltered  \\\n",
      "0                          2001.0                 77061   \n",
      "\n",
      "   all_reads_unfiltered_collapsed  \n",
      "0                            1996  \n",
      "running row 3\n",
      "this is for pipeline2 mCherryFWD, row3 in sheet \n",
      "\n",
      "minimum length after trimming is: 35\n",
      "tn5 analysis\n",
      "tn5 analysis\n",
      "sam file deleted\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N704 _ N501\n",
      "doing analysis on the global alignment N704 _ N501\n",
      "   %_target_site_reads  %_target_vector_reads  %_translocations  \\\n",
      "0            39.548755              27.870772         32.580473   \n",
      "\n",
      "   %_target_site_reads_collapsed  %_target_vector_reads_collapsed  \\\n",
      "0                      36.456064                        34.023239   \n",
      "\n",
      "   %_translocations_collapsed  alignments_target_site_all  \\\n",
      "0                   29.520697                       50325   \n",
      "\n",
      "   alignments_target_vector_all  alignments_translocations_all  \\\n",
      "0                         35465                          41458   \n",
      "\n",
      "   collapsed_alignments_target_site  collapsed_alignments_target_vector  \\\n",
      "0                              1004                                 937   \n",
      "\n",
      "   collapsed_alignments_translocations  total aligned reads filtered  \\\n",
      "0                                  813                      127248.0   \n",
      "\n",
      "   total collpased reads filtered  all_reads_unfiltered  \\\n",
      "0                          2754.0                147063   \n",
      "\n",
      "   all_reads_unfiltered_collapsed  \n",
      "0                            2765  \n"
     ]
    }
   ],
   "source": [
    "### summary function that analyzes the numbers in list. It puts together everything in following cells\n",
    "\n",
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "assembly_plus_targetvector = amplicon_info['genome_plus_targeting']\n",
    "file_genome_2bit_plus_target = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly_plus_targetvector + '.2bit')\n",
    "print(file_genome_2bit_plus_target)\n",
    "#Needs to have a bowtie build in which the targeting vector is included!!\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "#check the env variables\n",
    "#%env\n",
    "\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "############# Directory ############\n",
    "#this is a minimal directory of only ~300 files for quick debugging\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5/10000Reads'\n",
    "\n",
    "directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5'\n",
    "\n",
    "print('the directory is:', directory)\n",
    "\n",
    "\n",
    "seq_to_breaksite = [['TGGTTTTATTTCACCCCATGATAGTAATTATATCACTTCTGATCTGTTAAGAATAGACCTTTTAAAAGTATTGGATAACTTAGAGATGAGACATCTTCAGTTACTCTGTTATTCACCTATTACTCCTTAGGTTACTTGTGAATAATTTTGTGTGGGTCA', 'this is for pipeline2 polb fwd (row 0 in sheet)'],\n",
    "                    ['ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG','this is for pipeline2 polb rev (row 1 in sheet)'],\n",
    "                    ['GTTAGCAGACTTCCTCTGCCCTCAATCTTTAAGAAAAAAAAAAGTCTAACAATGATTTAGGAATGCTTTGAGGACTTAAATGATCTTATTGGAAACATACCAGTCTGCTAAAAGACTAATTTTGTGTGGGTCA','this is for pipeline2 mCherryRev, row2 in sheet'],\n",
    "                    ['GCAGGAGCTCGTCGACCCATGGGGGCCCGCCCCAACTGGGGTAACCTTTGGGCTCCCCGGGCGCGACTAGTGAATTCAGATCTGATATCTCTAGAAGTCCTGGG', 'this is for pipeline2 mCherryFWD, row3 in sheet'],\n",
    "                    ['TGGTTTTATTTCACCCCATGATAGTAATTATATCACTTCTGATCTGTTAAGAATAGACCTTTTAAAAGTATTGGATAACTTAGAGATGAGACATCTTCAGTTACTCTGTTATTCACCTATTACTCCTTAGGTTACTTGTGAATAATTTTGTGTGGGTCA', 'this is for pipeline2 WT polb fwd (row 4 in sheet)'],\n",
    "                    ['ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG','this is for pipeline2 WTpo lb rev (row 5 in sheet)']]\n",
    "\n",
    "\n",
    "# type the number in the list for the runs to run: \n",
    "    #0 is polbRev\n",
    "    #1 is polbRev \n",
    "    #2 is mcherry_rev\n",
    "    #3 is mcherry_fwd\n",
    "\n",
    "lines_to_run = [0,1,2,3]\n",
    "\n",
    "#type in 'on' or 'off'\n",
    "trim = 'off'\n",
    "local_align = 'off'\n",
    "global_align = 'on'\n",
    "\n",
    "for i in lines_to_run:\n",
    "    print('running row', i)\n",
    "    \n",
    "    #for each round choose the corresponding seq to breaksite\n",
    "    seq_primer_to_breaksite = seq_to_breaksite[i][0]\n",
    "    print(seq_to_breaksite[i][1],'\\n')\n",
    "    \n",
    "    #make the correct vector of the csv sheet\n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    \n",
    "    #run trimming\n",
    "    if trim == 'on':\n",
    "        trim_fastq_to_break(directory, amplicon_info, seq_primer_to_breaksite, length = 35)\n",
    "\n",
    "    if global_align == 'on':\n",
    "        #####this aligns global (end-to-end).  keep_sam=1 means keep the sam file\n",
    "        align_afterbreak_end_to_end_genome_global(directory, amplicon_info, assembly_plus_targetvector, keep_sam=0)\n",
    "\n",
    "\n",
    "        #### Process the alignment files to make a bam file of only high quality reads\n",
    "        final_trimmed_bam_filtered_mapq_AS_primary(directory, 'global', amplicon_info, 'tn5')\n",
    "\n",
    "        ### global alignment bed generation  \n",
    "        print('-----  now making bed file   ------')\n",
    "\n",
    "        N7 = amplicon_info['index_I1']\n",
    "        N5 = amplicon_info['index_I2']\n",
    "        file_trimmed_genome_global_bed = create_filename(directory, N7, N5, 'break_trimmed_genome_global_bed')\n",
    "        file_sorted_bam_genome_global = create_filename(directory, N7, N5, 'break_trimmed_filtered_and_sorted_genome_global')\n",
    "\n",
    "        !bam2bed < {file_sorted_bam_genome_global} > {file_trimmed_genome_global_bed}\n",
    "        bed_folder = os.path.abspath(create_filename(directory, N7, N5, 'all_bed'))\n",
    "\n",
    "        if not os.path.exists(bed_folder):\n",
    "            os.mkdir(bed_folder)\n",
    "        shutil.copy(file_trimmed_genome_global_bed, bed_folder)\n",
    "        print('made global bed file:', N7, '_', N5 )\n",
    "        \n",
    "        #quantify the global alignment\n",
    "        print('doing analysis on the global alignment', N7, '_', N5 )\n",
    "        quantify_global = quantify_pipeline2_alignments(directory, 'global', amplicon_info)\n",
    "        print(quantify_global)\n",
    "        \n",
    "    if local_align == 'on':\n",
    "        \n",
    "        # aligns local (soft clip the ends) keep_sam=1 means keep the sam file\n",
    "        align_afterbreaks_genome_local(directory, 1, amplicon_info, assembly_plus_targetvector, keep_sam=0)\n",
    "        \n",
    "        #sort out the good alignments\n",
    "        final_trimmed_bam_filtered_mapq_AS_primary(directory, 'local', amplicon_info, 'tn5')\n",
    "\n",
    "        # making a bed file for local\n",
    "        print('-----  now making bed file   ------')\n",
    "        N7 = amplicon_info['index_I1']\n",
    "        N5 = amplicon_info['index_I2']\n",
    "        file_trimmed_genome_local_bed = create_filename(directory, N7, N5, 'break_trimmed_genome_local_bed')\n",
    "        file_sorted_bam_genome_local = create_filename(directory, N7, N5, 'break_trimmed_filtered_and_sorted_bam_genome_local')\n",
    "\n",
    "        !bam2bed < {file_sorted_bam_genome_local} > {file_trimmed_genome_local_bed}\n",
    "        bed_folder = os.path.abspath(create_filename(directory, N7, N5, 'all_bed'))\n",
    "\n",
    "        if not os.path.exists(bed_folder):\n",
    "            os.mkdir(bed_folder)\n",
    "        shutil.copy(file_trimmed_genome_local_bed, bed_folder)\n",
    "        print('made local bed file:', N7, '_', N5 )\n",
    "        \n",
    "        # quantify the local alignment\n",
    "        print('doing the quantification of local alignments')\n",
    "        quantify_local = quantify_pipeline2_alignments(directory, 'local', amplicon_info, 'lam')\n",
    "        print(quantify_local)\n",
    "   \n",
    "    \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## this is if you want to do pipeline2 one piece at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5\n",
      "\n",
      "this is for pipeline2 polb fwd (row 0 in sheet) \n",
      "\n",
      "seq_to_breaksite TGGTTTTATTTCACCCCATGATAGTAATTATATCACTTCTGATCTGTTAAGAATAGACCTTTTAAAAGTATTGGATAACTTAGAGATGAGACATCTTCAGTTACTCTGTTATTCACCTATTACTCCTTAGGTTACTTGTGAATAATTTTGTGTGGGTCA \n",
      "\n",
      "directory /media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                        Tn5_Pol50_pbF\n",
       "Sample                                                                               Polb\n",
       "description                                   MiniSeq_K562_PolbHIROS_50per_unsorted_polbF\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                      prE368\n",
       "I7_Index_ID                                                          P7_N701_SBS12nextera\n",
       "index_I1                                                                             N701\n",
       "barcode_I1                                                                       TAAGGCGA\n",
       "I5_Index_ID                                                              P5_UMI_N50_Tn5-A\n",
       "index_I2                                                                             N501\n",
       "barcode_I2                                                                       GCGATCTA\n",
       "primer1_ForNest                                                                     prE86\n",
       "primer                                                                             prE361\n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                            42349731\n",
       "end                                                                              42349755\n",
       "strand                                                                                  +\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "genome_plus_targeting                                              hg38_plus_targetvector\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                            pE049_WO_redundant\n",
       "plasmid_sequence                        ttacataacttacggtaaatggcccgcctggctgaccgcccaacga...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "########### all inputs required for pipeline2 #####\n",
    "#This is for the sequence to the breaksite for the pipeline of genomic alignment \n",
    "# make sure the sequences are capital\n",
    "\n",
    "# number in list of \n",
    "\n",
    "#0 is polbRev\n",
    "#1 is polbRev \n",
    "#2 is mcherry_rev\n",
    "#3 is mcherry_fwd\n",
    "\n",
    "\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off\n",
    "\n",
    "############# Directory ############\n",
    "#this is a minimal directory of only ~300 files for quick debugging\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5/10000Reads'\n",
    "\n",
    "directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/P_Eric4_Tn5'\n",
    "\n",
    "print(directory)\n",
    "\n",
    "\n",
    "print()\n",
    "seq_to_breaksite = [['TGGTTTTATTTCACCCCATGATAGTAATTATATCACTTCTGATCTGTTAAGAATAGACCTTTTAAAAGTATTGGATAACTTAGAGATGAGACATCTTCAGTTACTCTGTTATTCACCTATTACTCCTTAGGTTACTTGTGAATAATTTTGTGTGGGTCA', 'this is for pipeline2 polb fwd (row 0 in sheet)'],\n",
    "                    ['ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG','this is for pipeline2 polb rev (row 1 in sheet)'],\n",
    "                    ['GTTAGCAGACTTCCTCTGCCCTCAATCTTTAAGAAAAAAAAAAGTCTAACAATGATTTAGGAATGCTTTGAGGACTTAAATGATCTTATTGGAAACATACCAGTCTGCTAAAAGACTAATTTTGTGTGGGTCA','this is for pipeline2 mCherryRev, row2 in sheet'],\n",
    "                    ['GCAGGAGCTCGTCGACCCATGGGGGCCCGCCCCAACTGGGGTAACCTTTGGGCTCCCCGGGCGCGACTAGTGAATTCAGATCTGATATCTCTAGAAGTCCTGGG', 'this is for pipeline2 mCherryFWD, row3 in sheet'],\n",
    "                    ['TGGTTTTATTTCACCCCATGATAGTAATTATATCACTTCTGATCTGTTAAGAATAGACCTTTTAAAAGTATTGGATAACTTAGAGATGAGACATCTTCAGTTACTCTGTTATTCACCTATTACTCCTTAGGTTACTTGTGAATAATTTTGTGTGGGTCA', 'this is for pipeline2 WT polb fwd (row 4 in sheet)'],\n",
    "                    ['ACAAAAGAGGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG','this is for pipeline2 WTpo lb rev (row 5 in sheet)']]\n",
    "\n",
    "  \n",
    "master_number = 0\n",
    "    \n",
    "    \n",
    "seq_primer_to_breaksite = seq_to_breaksite[master_number][0]\n",
    "print(seq_to_breaksite[master_number][1],'\\n')    \n",
    "print('seq_to_breaksite', seq_primer_to_breaksite, '\\n')\n",
    "\n",
    "#check amplicon info and directory\n",
    "print('directory', directory)\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "\n",
    "#the amplicon info is related to the line on the csv file. It is indexed from 0. For PolbF we use 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "\n",
    "amplicon_info = get_csv_data(directory, master_number)\n",
    "\n",
    "###########check waht is printed below #######\n",
    "amplicon_info\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "minimum length after trimming is: 35\n",
      "tn5 analysis\n"
     ]
    }
   ],
   "source": [
    "#Trim the break up to the cut site\n",
    "\n",
    "trim_fastq_to_break(directory, amplicon_info, seq_primer_to_breaksite, length = 35)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### alignments of sequences trimmed to the break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tn5 analysis\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N701 _ N501\n"
     ]
    }
   ],
   "source": [
    "#this aligns everything after the break to end-to-end keep_sam=1 means keep the sam file\n",
    "align_afterbreak_end_to_end_genome_global(directory, amplicon_info, assembly_plus_targetvector, keep_sam=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tn5 analysis\n",
      "sam file deleted\n",
      "-----  now making bed file   ------\n",
      "made local bed file: N704 _ N501\n"
     ]
    }
   ],
   "source": [
    "#this aligns everything in the local format wehre it can soft clip the ends keep_sam=1 means keep the sam file\n",
    "align_afterbreaks_genome_local(directory, 1, amplicon_info, assembly_plus_targetvector, keep_sam=0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparing the files for plotting in R. This make the bed files\n",
    "#### The bed files are filtered for primary reads of high quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N701 _ N501\n"
     ]
    }
   ],
   "source": [
    "#this filteres all the GLOBAL reads.\n",
    "\n",
    "final_trimmed_bam_filtered_mapq_AS_primary(directory, 'global', amplicon_info)\n",
    "\n",
    "\n",
    "# THE BED IS MADE FROM THE FILTERED READS\n",
    "### This should work for automatic bed generation, but it doesn't produce anythign and no ideaw way\n",
    "\n",
    "print('-----  now making bed file   ------')\n",
    "\n",
    "\n",
    "N7 = amplicon_info['index_I1']\n",
    "N5 = amplicon_info['index_I2']\n",
    "file_trimmed_genome_global_bed = create_filename(directory, N7, N5, 'break_trimmed_genome_global_bed')\n",
    "file_sorted_bam_genome_global = create_filename(directory, N7, N5, 'break_trimmed_filtered_and_sorted_genome_global')\n",
    "\n",
    "!bam2bed < {file_sorted_bam_genome_global} > {file_trimmed_genome_global_bed}\n",
    "\n",
    "\n",
    "bed_folder = os.path.abspath(create_filename(directory, N7, N5, 'all_bed'))\n",
    "\n",
    "\n",
    "if not os.path.exists(bed_folder):\n",
    "    os.mkdir(bed_folder)\n",
    "\n",
    "shutil.copy(file_trimmed_genome_global_bed, bed_folder)\n",
    "\n",
    "print('made global bed file:', N7, '_', N5 )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering and making bed file for local alignments.  BED FILE IS MADE OF FILTERED READS\n",
    "\n",
    "final_trimmed_bam_filtered_mapq_AS_primary(directory, 'local', amplicon_info)\n",
    "\n",
    "\n",
    "print('-----  now making bed file   ------')\n",
    "\n",
    "N7 = amplicon_info['index_I1']\n",
    "N5 = amplicon_info['index_I2']\n",
    "file_trimmed_genome_local_bed = create_filename(directory, N7, N5, 'break_trimmed_genome_local_bed')\n",
    "file_sorted_bam_genome_local = create_filename(directory, N7, N5, 'break_trimmed_filtered_and_sorted_bam_genome_local')\n",
    "\n",
    "\n",
    "!bam2bed < {file_sorted_bam_genome_local} > {file_trimmed_genome_local_bed}\n",
    "\n",
    "\n",
    "bed_folder = os.path.abspath(create_filename(directory, N7, N5, 'all_bed'))\n",
    "\n",
    "if not os.path.exists(bed_folder):\n",
    "    os.mkdir(bed_folder)\n",
    "\n",
    "shutil.copy(file_trimmed_genome_local_bed, bed_folder)\n",
    "\n",
    "print('made local bed file:', N7, '_', N5 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### take the bed files and then use R script to plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>%_target_site_reads</th>\n",
       "      <th>%_target_vector_reads</th>\n",
       "      <th>%_translocations</th>\n",
       "      <th>%_target_site_reads_collapsed</th>\n",
       "      <th>%_target_vector_reads_collapsed</th>\n",
       "      <th>%_translocations_collapsed</th>\n",
       "      <th>alignments_target_site_all</th>\n",
       "      <th>alignments_target_vector_all</th>\n",
       "      <th>alignments_translocations_all</th>\n",
       "      <th>collapsed_alignments_target_site</th>\n",
       "      <th>collapsed_alignments_target_vector</th>\n",
       "      <th>collapsed_alignments_translocations</th>\n",
       "      <th>total aligned reads</th>\n",
       "      <th>total collpased reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>70.714524</td>\n",
       "      <td>28.727636</td>\n",
       "      <td>0.55784</td>\n",
       "      <td>71.733871</td>\n",
       "      <td>28.104839</td>\n",
       "      <td>0.16129</td>\n",
       "      <td>127272</td>\n",
       "      <td>51704</td>\n",
       "      <td>1004</td>\n",
       "      <td>1779</td>\n",
       "      <td>697</td>\n",
       "      <td>4</td>\n",
       "      <td>179980.0</td>\n",
       "      <td>2480.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   %_target_site_reads  %_target_vector_reads  %_translocations  \\\n",
       "0            70.714524              28.727636           0.55784   \n",
       "\n",
       "   %_target_site_reads_collapsed  %_target_vector_reads_collapsed  \\\n",
       "0                      71.733871                        28.104839   \n",
       "\n",
       "   %_translocations_collapsed  alignments_target_site_all  \\\n",
       "0                     0.16129                      127272   \n",
       "\n",
       "   alignments_target_vector_all  alignments_translocations_all  \\\n",
       "0                         51704                           1004   \n",
       "\n",
       "   collapsed_alignments_target_site  collapsed_alignments_target_vector  \\\n",
       "0                              1779                                 697   \n",
       "\n",
       "   collapsed_alignments_translocations  total aligned reads  \\\n",
       "0                                    4             179980.0   \n",
       "\n",
       "   total collpased reads  \n",
       "0                 2480.0  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#polb fwd\n",
    "print(amplicon_info['name'])\n",
    "quantify_pipeline2_alignments(directory, 'global', amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alignments_target_site_all</th>\n",
       "      <th>alignments_target_vector_all</th>\n",
       "      <th>alignments_translocations_all</th>\n",
       "      <th>collapsed_alignments_target_site</th>\n",
       "      <th>collapsed_alignments_target_vector</th>\n",
       "      <th>collapsed_alignments_translocations</th>\n",
       "      <th>total aligned reads</th>\n",
       "      <th>total collpased reads</th>\n",
       "      <th>%_target_site_reads</th>\n",
       "      <th>%_target_vector_reads</th>\n",
       "      <th>%_translocations</th>\n",
       "      <th>%_target_site_reads_collapsed</th>\n",
       "      <th>%_target_vector_reads_collapsed</th>\n",
       "      <th>%_translocations_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>44362</td>\n",
       "      <td>53745</td>\n",
       "      <td>200</td>\n",
       "      <td>1094</td>\n",
       "      <td>1062</td>\n",
       "      <td>138</td>\n",
       "      <td>98307.0</td>\n",
       "      <td>2294.0</td>\n",
       "      <td>45.125983</td>\n",
       "      <td>54.670573</td>\n",
       "      <td>0.203444</td>\n",
       "      <td>47.689625</td>\n",
       "      <td>46.294682</td>\n",
       "      <td>6.015693</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   alignments_target_site_all  alignments_target_vector_all  \\\n",
       "0                       44362                         53745   \n",
       "\n",
       "   alignments_translocations_all  collapsed_alignments_target_site  \\\n",
       "0                            200                              1094   \n",
       "\n",
       "   collapsed_alignments_target_vector  collapsed_alignments_translocations  \\\n",
       "0                                1062                                  138   \n",
       "\n",
       "   total aligned reads  total collpased reads  %_target_site_reads  \\\n",
       "0              98307.0                 2294.0            45.125983   \n",
       "\n",
       "   %_target_vector_reads  %_translocations  %_target_site_reads_collapsed  \\\n",
       "0              54.670573          0.203444                      47.689625   \n",
       "\n",
       "   %_target_vector_reads_collapsed  %_translocations_collapsed  \n",
       "0                        46.294682                    6.015693  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#polb rev\n",
    "print(amplicon_info['name'])\n",
    "quantify_pipeline2_alignments(directory, 'global', amplicon_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alignments_target_site_all</th>\n",
       "      <th>alignments_target_vector_all</th>\n",
       "      <th>alignments_translocations_all</th>\n",
       "      <th>collapsed_alignments_target_site</th>\n",
       "      <th>collapsed_alignments_target_vector</th>\n",
       "      <th>collapsed_alignments_translocations</th>\n",
       "      <th>total aligned reads</th>\n",
       "      <th>total collpased reads</th>\n",
       "      <th>%_target_site_reads</th>\n",
       "      <th>%_target_vector_reads</th>\n",
       "      <th>%_translocations</th>\n",
       "      <th>%_target_site_reads_collapsed</th>\n",
       "      <th>%_target_vector_reads_collapsed</th>\n",
       "      <th>%_translocations_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>50325</td>\n",
       "      <td>35465</td>\n",
       "      <td>41458</td>\n",
       "      <td>1004</td>\n",
       "      <td>937</td>\n",
       "      <td>813</td>\n",
       "      <td>127248.0</td>\n",
       "      <td>2754.0</td>\n",
       "      <td>39.548755</td>\n",
       "      <td>27.870772</td>\n",
       "      <td>32.580473</td>\n",
       "      <td>36.456064</td>\n",
       "      <td>34.023239</td>\n",
       "      <td>29.520697</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   alignments_target_site_all  alignments_target_vector_all  \\\n",
       "0                       50325                         35465   \n",
       "\n",
       "   alignments_translocations_all  collapsed_alignments_target_site  \\\n",
       "0                          41458                              1004   \n",
       "\n",
       "   collapsed_alignments_target_vector  collapsed_alignments_translocations  \\\n",
       "0                                 937                                  813   \n",
       "\n",
       "   total aligned reads  total collpased reads  %_target_site_reads  \\\n",
       "0             127248.0                 2754.0            39.548755   \n",
       "\n",
       "   %_target_vector_reads  %_translocations  %_target_site_reads_collapsed  \\\n",
       "0              27.870772         32.580473                      36.456064   \n",
       "\n",
       "   %_target_vector_reads_collapsed  %_translocations_collapsed  \n",
       "0                        34.023239                   29.520697  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#mcherry fwd\n",
    "print(amplicon_info['name'])\n",
    "quantify_pipeline2_alignments(directory, 'global', amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alignments_target_site_all</th>\n",
       "      <th>alignments_target_vector_all</th>\n",
       "      <th>alignments_translocations_all</th>\n",
       "      <th>collapsed_alignments_target_site</th>\n",
       "      <th>collapsed_alignments_target_vector</th>\n",
       "      <th>collapsed_alignments_translocations</th>\n",
       "      <th>total aligned reads</th>\n",
       "      <th>total collpased reads</th>\n",
       "      <th>%_target_site_reads</th>\n",
       "      <th>%_target_vector_reads</th>\n",
       "      <th>%_translocations</th>\n",
       "      <th>%_target_site_reads_collapsed</th>\n",
       "      <th>%_target_vector_reads_collapsed</th>\n",
       "      <th>%_translocations_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17974</td>\n",
       "      <td>40909</td>\n",
       "      <td>12349</td>\n",
       "      <td>493</td>\n",
       "      <td>1157</td>\n",
       "      <td>351</td>\n",
       "      <td>71232.0</td>\n",
       "      <td>2001.0</td>\n",
       "      <td>25.233041</td>\n",
       "      <td>57.430649</td>\n",
       "      <td>17.33631</td>\n",
       "      <td>24.637681</td>\n",
       "      <td>57.821089</td>\n",
       "      <td>17.541229</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   alignments_target_site_all  alignments_target_vector_all  \\\n",
       "0                       17974                         40909   \n",
       "\n",
       "   alignments_translocations_all  collapsed_alignments_target_site  \\\n",
       "0                          12349                               493   \n",
       "\n",
       "   collapsed_alignments_target_vector  collapsed_alignments_translocations  \\\n",
       "0                                1157                                  351   \n",
       "\n",
       "   total aligned reads  total collpased reads  %_target_site_reads  \\\n",
       "0              71232.0                 2001.0            25.233041   \n",
       "\n",
       "   %_target_vector_reads  %_translocations  %_target_site_reads_collapsed  \\\n",
       "0              57.430649          17.33631                      24.637681   \n",
       "\n",
       "   %_target_vector_reads_collapsed  %_translocations_collapsed  \n",
       "0                        57.821089                   17.541229  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#mcherry rev\n",
    "print(amplicon_info['name'])\n",
    "quantify_pipeline2_alignments(directory, 'global', amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
