{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "### File Data\n",
    "#sorted but not anlayzed files: /media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed/\n",
    "#Take the first 500 files (2000 lines)\n",
    "\n",
    "#### of all 4 file types - If you do it multiple times it just appends text lines to bottom of text editor\n",
    "#(terminal)\n",
    "#$head -n2000 N706_N504_R1.fastq >> ./10000Reads/N706_N504_R1.fastq\n",
    "#$head -n2000 N706_N504_R2.fastq >> ./10000Reads/N706_N504_R2.fastq\n",
    "\n",
    "#$head -n2000 N707_N505_R1.fastq >> ./10000Reads/N707_N505_R1.fastq\n",
    "#$head -n2000 N707_N505_R2.fastq >> ./10000Reads/N707_N505_R2.fastq\n",
    "\n",
    "#### check how many lines: (terminal)\n",
    "#$wc -l N706_N504_R1.fastq\n",
    "\n",
    "### can compress with gZip\n",
    "#$gzip N706_N505_R1.fastq\n",
    "#$gzip N706_N505_R2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "\n",
    "from uditas.uditas_replace_helpers import *\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Pipeline1\n",
    "# This, for the most part, is the UDITAS pipeline modified to allow for REPLACE targeting. \n",
    "\n",
    "## It is not possible to have UMI counts through LAM. So that info is not real. Ignore it.\n",
    "## There is a second pipeline below to do global alignments to quantify large deletions, translocatins, concatomers, etc by aligning the sequence past the break site to the entire genome+targeting sequence.\n",
    "\n",
    "### 1\n",
    "\n",
    "\n",
    "Should also note that this is slightly different than the Tn5. It needs clipping of more sequence. We create false UMIS to satisfy the UDITAS pipeline by taking the first 5nt of the sequence. Also the adapter needs to be clipped. So a new function was made. Also the direction of read1:read2 is switched in order because of our choice of p5 and p7 attached to the gene specific sequence.\n",
    "\n",
    "\n",
    "## Inputs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "#things taken from original uditas args\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM/10000Reads\n"
     ]
    }
   ],
   "source": [
    "#Directory\n",
    "\n",
    "#this is a minimal directory of only 300 files for quick debugging\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM'\n",
    "\n",
    "directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM/10000Reads'\n",
    "\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this is for pipeline1 LAM mCherryfwd N707 N505, row2 in sheet \n",
      "\n",
      "primer_seq_plus_downstream TGCATCGCATTGTCTGAGTAGGTGTCATTCTATTC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                        LAM_Pol50_bpA\n",
       "Sample                                                                               Polb\n",
       "description                                         MiniSeq_K562_Polb_mCherrFwd_50per_LAM\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                         NaN\n",
       "I7_Index_ID                                                          P7_N707_SBS12nextera\n",
       "index_I1                                                                             N707\n",
       "barcode_I1                                                                       CTCTCTAC\n",
       "I5_Index_ID                                                             P5_UMI_N505_Tn5-A\n",
       "index_I2                                                                             N505\n",
       "barcode_I2                                                                       CTCCTTAC\n",
       "primer_outfwd                                                                         NaN\n",
       "primer                                                                U1-C-bpA_Fwd_nest2 \n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                            42350359\n",
       "end                                                                              42350378\n",
       "strand                                                                                  -\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "genome_plus_targeting                                              hg38_plus_targetvector\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                            pE049_WO_redundant\n",
       "plasmid_sequence                        ttacataacttacggtaaatggcccgcctggctgaccgcccaacga...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 2, dtype: object"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number in list of which sheet to access in the sample_info.csv\n",
    "\n",
    "#0 is LAM polb rev WT N706 N504 (row 0 in sheet)\n",
    "#1 is LAM polb rev Targeting 50% N706 N505(row 1 in sheet)\n",
    "#2 is LAM mCherryRev N707 N505, row2 in sheet'\n",
    "#3 is LAM mCherryFWD   N708 N505 , row3 in sheet'\n",
    "\n",
    "\n",
    "master_number = 2\n",
    "\n",
    "\n",
    "#################################################################\n",
    "#primer sequences for mispriming. Primer sequence + 12nt downstream \n",
    "#this will only be avialable for read1  (different than for Tn5 setup)\n",
    "\n",
    "##### the sequences need to be capital for this to work########\n",
    "\n",
    "\n",
    "primerlist = [['GGCCAAGCTGGAGCAGGAAATAGATGCACACG', 'this is for pipeline1 LAM- polb rev WT N706 N504 (row 0 in sheet)'],\n",
    "              ['GGCCAAGCTGGAGCAGGAAATAGATGCACACG', 'this is for pipeline1 LAM polb rev Targeting 50% N706 N505(row 1 in sheet)'],\n",
    "              ['TGCATCGCATTGTCTGAGTAGGTGTCATTCTATTC', 'this is for pipeline1 LAM mCherryfwd N707 N505, row2 in sheet'],\n",
    "              ['CCTCCACGTCACCGCATGTTAGCAGACTTCCT', 'this is for mCherryRev LAM  N708 N505 , row3 in sheet']]\n",
    "\n",
    "                \n",
    "\n",
    "primer_seq_plus_downstream = primerlist[master_number][0]          \n",
    "print(primerlist[master_number][1], '\\n')    \n",
    "print('primer_seq_plus_downstream', primer_seq_plus_downstream)\n",
    "\n",
    "\n",
    "#check what is printed below!!!\n",
    "\n",
    "########################################################################\n",
    "# Function to make the 'amplicon_info' list. Taking the line of our experiment csv file\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "\n",
    "#the amplicon info is related to the line on the csv file. It is indexed from 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "\n",
    "amplicon_info = get_csv_data(directory, master_number)\n",
    "\n",
    "###########check waht is printed below #######\n",
    "amplicon_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "assembly = amplicon_info['genome']\n",
    "file_genome_2bit = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly + '.2bit')\n",
    "print(file_genome_2bit)\n",
    "\n",
    "# BOWTIE2_INDEXES are needed for global alignments\n",
    "#not sure if this will work\n",
    "#normally in bash: export BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pull out the UMI info and clip the adapter sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reads: 500\n",
      "number of unique umis 500\n"
     ]
    }
   ],
   "source": [
    "umi_joining(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discard Mispriming Reads\n",
    "When you put a universal primer on the ends of everything, every mispriming event will amplify. An effect we normally don't deal with. I did nested PCR to reduce this. However 85% of the alignments in the UDITAS data I looked at seemed to be mispriming. They did all their blasting and analysis before removing mispriming. But to save computational power and remove error early on I will discard mispriming events. \n",
    "They discard these only for plasmid alignments analyze_alignments_plasmid for some reason which comes from the bam file.\n",
    "\n",
    "### read 1 and 2 are switched. By that I mean the gene specific primer is attached to the P7 sequence instead of P5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total_read_number 500\n",
      "reads_with_good_priming 470\n",
      "reads_with_bad_priming 30\n",
      "perfect matched 444\n",
      "good but mismatched 26\n"
     ]
    }
   ],
   "source": [
    "correct_priming_lam(directory, amplicon_info, primer_seq_plus_downstream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Running the trimming\n",
    "\n",
    "trim_fastq(directory, amplicon_info, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running the reference plasmid creation\n",
    "create_plasmid_reference(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'replace'"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Test the get_reaction_type is working\n",
    "get_reaction_type(amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make the amplicons by calling the create_amplicon function\n",
    "create_amplicon(directory, amplicon_info, file_genome_2bit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try out the alignment to the plasmid\n",
    "align_plasmid_local(directory, amplicon_info, ncpu=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract the unmapped reads\n",
    "extract_unmapped_reads_plasmid(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target_plus_plasmid_total_reads  target_plus_plasmid_total_reads_collapsed  \\\n",
       "0                                0                                          0   \n",
       "\n",
       "   plasmid_only_total_reads  plasmid_only_total_reads_collapsed  \n",
       "0                         2                                   2  "
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#run the plasmid analysis to coudn plasmid integration events\n",
    "result_plasmid_df = analyze_alignments_plasmid(directory, amplicon_info, min_MAPQ, file_genome_2bit, True)\n",
    "result_plasmid_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "#align against our suite of amplicons\n",
    "align_amplicon(directory, amplicon_info, check_plasmid_insertions, ncpu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this will extract unmap reads new folder (files that did not align to the predicted structural variants)\n",
    "extract_unmapped_reads_amplicons(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This section is to do the analysis of the amplicon alignments\n",
    "\n",
    "##### need to check my UMI stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reads before mispriming cleanup 500\n",
      "number of unique umis before mispriming cleanup 500\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   all_amplicons_total_reads  all_amplicons_total_reads_collapsed\n",
       "0                        145                                  145"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_reads_in_all_amplicons_df = analyze_alignments_all_amplicons(directory, amplicon_info, min_MAPQ, min_AS)\n",
    "result_reads_in_all_amplicons_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_cut1_total_reads</th>\n",
       "      <th>wt_cut1_total_indels</th>\n",
       "      <th>wt_cut1_total_deletions</th>\n",
       "      <th>wt_cut1_total_insertions</th>\n",
       "      <th>wt_cut1_total_reads_collapsed</th>\n",
       "      <th>wt_cut1_total_indels_collapsed</th>\n",
       "      <th>wt_cut1_total_deletions_collapsed</th>\n",
       "      <th>wt_cut1_total_insertions_collapsed</th>\n",
       "      <th>wt_cut2_total_reads</th>\n",
       "      <th>wt_cut2_total_indels</th>\n",
       "      <th>...</th>\n",
       "      <th>1a_1a_cut1_total_insertions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_reads</th>\n",
       "      <th>2b_2b_cut1_total_indels</th>\n",
       "      <th>2b_2b_cut1_total_deletions</th>\n",
       "      <th>2b_2b_cut1_total_insertions</th>\n",
       "      <th>2b_2b_cut1_total_reads_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_indels_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_deletions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_insertions_collapsed</th>\n",
       "      <th>median_fragment_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>320.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 113 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   wt_cut1_total_reads  wt_cut1_total_indels  wt_cut1_total_deletions  \\\n",
       "0                    0                     0                        0   \n",
       "\n",
       "   wt_cut1_total_insertions  wt_cut1_total_reads_collapsed  \\\n",
       "0                         0                              0   \n",
       "\n",
       "   wt_cut1_total_indels_collapsed  wt_cut1_total_deletions_collapsed  \\\n",
       "0                               0                                  0   \n",
       "\n",
       "   wt_cut1_total_insertions_collapsed  wt_cut2_total_reads  \\\n",
       "0                                   0                    0   \n",
       "\n",
       "   wt_cut2_total_indels  ...  1a_1a_cut1_total_insertions_collapsed  \\\n",
       "0                     0  ...                                      0   \n",
       "\n",
       "   2b_2b_cut1_total_reads  2b_2b_cut1_total_indels  \\\n",
       "0                       0                        0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions  2b_2b_cut1_total_insertions  \\\n",
       "0                           0                            0   \n",
       "\n",
       "   2b_2b_cut1_total_reads_collapsed  2b_2b_cut1_total_indels_collapsed  \\\n",
       "0                                 0                                  0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions_collapsed  \\\n",
       "0                                     0   \n",
       "\n",
       "   2b_2b_cut1_total_insertions_collapsed  median_fragment_size  \n",
       "0                                      0                 320.0  \n",
       "\n",
       "[1 rows x 113 columns]"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_amplicon_df = analyze_alignments(directory, amplicon_info, window_size, amplicon_window_around_cut, min_MAPQ, min_AS)\n",
    "result_amplicon_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is the Global alignment of the remaining reads\n",
    "\n",
    "#### What I need here is to modify it so that "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run the allignment\n",
    "assembly = amplicon_info['genome']\n",
    "align_genome_global(directory, amplicon_info, assembly)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of alignments that were attempted to align 4\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genomewide_total_reads  genomewide_total_reads_collapsed  \\\n",
       "0                       2                                 2   \n",
       "\n",
       "   genomewide_target_only_reads  genomewide_target_only_reads_collapsed  \n",
       "0                             0                                       0  "
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Do the global alignment analysis\n",
    "results_genome_global_df = analyze_alignments_genome_global(directory, amplicon_info, min_MAPQ, min_AS,  file_genome_2bit)\n",
    "results_genome_global_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summerize data Pipeline1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "470.0"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_count = pd.DataFrame()\n",
    "read_count.loc[0,'read_count'] = count_reads(directory, amplicon_info)\n",
    "read_count.loc[0,'read_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_count</th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "      <th>total_aligned</th>\n",
       "      <th>total_aligned_collapsed</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>percent_aligned_all_amplicons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>470.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>145</td>\n",
       "      <td>145</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "      <td>31.702128</td>\n",
       "      <td>97.315436</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   read_count  target_plus_plasmid_total_reads  \\\n",
       "0       470.0                                0   \n",
       "\n",
       "   target_plus_plasmid_total_reads_collapsed  plasmid_only_total_reads  \\\n",
       "0                                          0                         2   \n",
       "\n",
       "   plasmid_only_total_reads_collapsed  all_amplicons_total_reads  \\\n",
       "0                                   2                        145   \n",
       "\n",
       "   all_amplicons_total_reads_collapsed  genomewide_total_reads  \\\n",
       "0                                  145                       2   \n",
       "\n",
       "   genomewide_total_reads_collapsed  genomewide_target_only_reads  \\\n",
       "0                                 2                             0   \n",
       "\n",
       "   genomewide_target_only_reads_collapsed  total_aligned  \\\n",
       "0                                       0            149   \n",
       "\n",
       "   total_aligned_collapsed  percent_aligned  percent_aligned_all_amplicons  \n",
       "0                      149        31.702128                      97.315436  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#summerize the data\n",
    "\n",
    "summary_all_alignments = get_summary_all_alignments(directory, amplicon_info, read_count.loc[[0]], result_plasmid_df, result_reads_in_all_amplicons_df, results_genome_global_df)\n",
    "summary_all_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NGS_req-ID</th>\n",
       "      <th>name</th>\n",
       "      <th>Sample</th>\n",
       "      <th>description</th>\n",
       "      <th>Control sample (Y/N)</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Dilution</th>\n",
       "      <th>Cell name_type</th>\n",
       "      <th>UMI_Len</th>\n",
       "      <th>IndexI7Primer</th>\n",
       "      <th>...</th>\n",
       "      <th>replace_rev_cut2_total_reads_collapsed_percent</th>\n",
       "      <th>doner_tail_tail_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>doner_head_tail_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>doner_head_head_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>1a_1a_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>2b_2b_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed_percent</th>\n",
       "      <th>plasmid_only_total_reads_collapsed_percent</th>\n",
       "      <th>Type</th>\n",
       "      <th>Percent Editing</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>large_inversion_cut2_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>66.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.508772</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>3.508772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows × 188 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     NGS_req-ID                name Sample  \\\n",
       "0    A000H2GWGF          LAM_WT_pbR     WT   \n",
       "1    A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "2    A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "3    A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "4    A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "5    A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "6    A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "7    A000H2GWGF          LAM_WT_pbR     WT   \n",
       "8    A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "9    A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "10   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "11   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "12   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "13   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "14   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "15   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "16   A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "17   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "18   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "19   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "20   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "21   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "22   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "23   A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "24   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "25   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "26   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "27   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "28   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "29   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "..          ...                 ...    ...   \n",
       "82   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "83   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "84   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "85   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "86   A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "87   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "88   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "89   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "90   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "91   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "92   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "93   A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "94   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "95   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "96   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "97   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "98   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "99   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "100  A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "101  A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "102  A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "103  A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "104  A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "105  A000H2GWGF          LAM_WT_pbR     WT   \n",
       "106  A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "107  A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "108  A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "109  A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "110  A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "111  A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "\n",
       "                               description Control sample (Y/N)  Notes  \\\n",
       "0             MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "1          MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "2    MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "3    MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "4          MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "5    MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "6    MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "7             MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "8          MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "9    MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "10   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "11         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "12   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "13   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "14            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "15         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "16   MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "17   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "18         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "19   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "20   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "21            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "22         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "23   MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "24   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "25         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "26   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "27   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "28            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "29         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "..                                     ...                  ...    ...   \n",
       "82   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "83   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "84            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "85         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "86   MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "87   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "88         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "89   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "90   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "91            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "92         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "93   MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "94   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "95         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "96   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "97   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "98            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "99         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "100  MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "101  MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "102        MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "103  MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "104  MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "105           MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "106        MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "107  MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "108  MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "109        MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "110  MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "111  MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "\n",
       "     Dilution Cell name_type     UMI_Len  IndexI7Primer  ...  \\\n",
       "0         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "1         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "2         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "3         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "4         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "5         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "6         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "7         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "8         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "9         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "10        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "11        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "12        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "13        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "14        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "15        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "16        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "17        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "18        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "19        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "20        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "21        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "22        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "23        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "24        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "25        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "26        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "27        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "28        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "29        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "..        ...            ...         ...            ...  ...   \n",
       "82        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "83        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "84        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "85        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "86        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "87        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "88        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "89        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "90        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "91        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "92        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "93        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "94        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "95        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "96        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "97        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "98        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "99        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "100       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "101       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "102       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "103       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "104       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "105       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "106       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "107       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "108       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "109       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "110       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "111       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "\n",
       "    replace_rev_cut2_total_reads_collapsed_percent  \\\n",
       "0                                              0.0   \n",
       "1                                              NaN   \n",
       "2                                              NaN   \n",
       "3                                              NaN   \n",
       "4                                              NaN   \n",
       "5                                              NaN   \n",
       "6                                              NaN   \n",
       "7                                              0.0   \n",
       "8                                              NaN   \n",
       "9                                              NaN   \n",
       "10                                             NaN   \n",
       "11                                             NaN   \n",
       "12                                             NaN   \n",
       "13                                             NaN   \n",
       "14                                             0.0   \n",
       "15                                             NaN   \n",
       "16                                             NaN   \n",
       "17                                             NaN   \n",
       "18                                             NaN   \n",
       "19                                             NaN   \n",
       "20                                             NaN   \n",
       "21                                             0.0   \n",
       "22                                             NaN   \n",
       "23                                             NaN   \n",
       "24                                             NaN   \n",
       "25                                             NaN   \n",
       "26                                             NaN   \n",
       "27                                             NaN   \n",
       "28                                             0.0   \n",
       "29                                             NaN   \n",
       "..                                             ...   \n",
       "82                                             NaN   \n",
       "83                                             NaN   \n",
       "84                                             0.0   \n",
       "85                                             NaN   \n",
       "86                                             NaN   \n",
       "87                                             NaN   \n",
       "88                                             NaN   \n",
       "89                                             NaN   \n",
       "90                                             NaN   \n",
       "91                                             0.0   \n",
       "92                                             NaN   \n",
       "93                                             NaN   \n",
       "94                                             NaN   \n",
       "95                                             NaN   \n",
       "96                                             NaN   \n",
       "97                                             NaN   \n",
       "98                                             0.0   \n",
       "99                                             NaN   \n",
       "100                                            NaN   \n",
       "101                                            NaN   \n",
       "102                                            NaN   \n",
       "103                                            NaN   \n",
       "104                                            NaN   \n",
       "105                                            0.0   \n",
       "106                                            NaN   \n",
       "107                                            NaN   \n",
       "108                                            NaN   \n",
       "109                                            NaN   \n",
       "110                                            NaN   \n",
       "111                                            NaN   \n",
       "\n",
       "    doner_tail_tail_cut1_total_reads_collapsed_percent  \\\n",
       "0                                                  0.0   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                                  0.0   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                                 0.0   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                                 0.0   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                                 0.0   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "82                                                 NaN   \n",
       "83                                                 NaN   \n",
       "84                                                 0.0   \n",
       "85                                                 NaN   \n",
       "86                                                 NaN   \n",
       "87                                                 NaN   \n",
       "88                                                 NaN   \n",
       "89                                                 NaN   \n",
       "90                                                 NaN   \n",
       "91                                                 0.0   \n",
       "92                                                 NaN   \n",
       "93                                                 NaN   \n",
       "94                                                 NaN   \n",
       "95                                                 NaN   \n",
       "96                                                 NaN   \n",
       "97                                                 NaN   \n",
       "98                                                 0.0   \n",
       "99                                                 NaN   \n",
       "100                                                NaN   \n",
       "101                                                NaN   \n",
       "102                                                NaN   \n",
       "103                                                NaN   \n",
       "104                                                NaN   \n",
       "105                                                0.0   \n",
       "106                                                NaN   \n",
       "107                                                NaN   \n",
       "108                                                NaN   \n",
       "109                                                NaN   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "\n",
       "    doner_head_tail_cut1_total_reads_collapsed_percent  \\\n",
       "0                                            66.666667   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                            66.666667   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                           66.666667   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                           66.666667   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                           66.666667   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "82                                                 NaN   \n",
       "83                                                 NaN   \n",
       "84                                           66.666667   \n",
       "85                                                 NaN   \n",
       "86                                                 NaN   \n",
       "87                                                 NaN   \n",
       "88                                                 NaN   \n",
       "89                                                 NaN   \n",
       "90                                                 NaN   \n",
       "91                                           66.666667   \n",
       "92                                                 NaN   \n",
       "93                                                 NaN   \n",
       "94                                                 NaN   \n",
       "95                                                 NaN   \n",
       "96                                                 NaN   \n",
       "97                                                 NaN   \n",
       "98                                           66.666667   \n",
       "99                                                 NaN   \n",
       "100                                                NaN   \n",
       "101                                                NaN   \n",
       "102                                                NaN   \n",
       "103                                                NaN   \n",
       "104                                                NaN   \n",
       "105                                          66.666667   \n",
       "106                                                NaN   \n",
       "107                                                NaN   \n",
       "108                                                NaN   \n",
       "109                                                NaN   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "\n",
       "    doner_head_head_cut1_total_reads_collapsed_percent  \\\n",
       "0                                                  0.0   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                                  0.0   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                                 0.0   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                                 0.0   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                                 0.0   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "82                                                 NaN   \n",
       "83                                                 NaN   \n",
       "84                                                 0.0   \n",
       "85                                                 NaN   \n",
       "86                                                 NaN   \n",
       "87                                                 NaN   \n",
       "88                                                 NaN   \n",
       "89                                                 NaN   \n",
       "90                                                 NaN   \n",
       "91                                                 0.0   \n",
       "92                                                 NaN   \n",
       "93                                                 NaN   \n",
       "94                                                 NaN   \n",
       "95                                                 NaN   \n",
       "96                                                 NaN   \n",
       "97                                                 NaN   \n",
       "98                                                 0.0   \n",
       "99                                                 NaN   \n",
       "100                                                NaN   \n",
       "101                                                NaN   \n",
       "102                                                NaN   \n",
       "103                                                NaN   \n",
       "104                                                NaN   \n",
       "105                                                0.0   \n",
       "106                                                NaN   \n",
       "107                                                NaN   \n",
       "108                                                NaN   \n",
       "109                                                NaN   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "\n",
       "    1a_1a_cut1_total_reads_collapsed_percent  \\\n",
       "0                                        0.0   \n",
       "1                                        NaN   \n",
       "2                                        NaN   \n",
       "3                                        NaN   \n",
       "4                                        NaN   \n",
       "5                                        NaN   \n",
       "6                                        NaN   \n",
       "7                                        0.0   \n",
       "8                                        NaN   \n",
       "9                                        NaN   \n",
       "10                                       NaN   \n",
       "11                                       NaN   \n",
       "12                                       NaN   \n",
       "13                                       NaN   \n",
       "14                                       0.0   \n",
       "15                                       NaN   \n",
       "16                                       NaN   \n",
       "17                                       NaN   \n",
       "18                                       NaN   \n",
       "19                                       NaN   \n",
       "20                                       NaN   \n",
       "21                                       0.0   \n",
       "22                                       NaN   \n",
       "23                                       NaN   \n",
       "24                                       NaN   \n",
       "25                                       NaN   \n",
       "26                                       NaN   \n",
       "27                                       NaN   \n",
       "28                                       0.0   \n",
       "29                                       NaN   \n",
       "..                                       ...   \n",
       "82                                       NaN   \n",
       "83                                       NaN   \n",
       "84                                       0.0   \n",
       "85                                       NaN   \n",
       "86                                       NaN   \n",
       "87                                       NaN   \n",
       "88                                       NaN   \n",
       "89                                       NaN   \n",
       "90                                       NaN   \n",
       "91                                       0.0   \n",
       "92                                       NaN   \n",
       "93                                       NaN   \n",
       "94                                       NaN   \n",
       "95                                       NaN   \n",
       "96                                       NaN   \n",
       "97                                       NaN   \n",
       "98                                       0.0   \n",
       "99                                       NaN   \n",
       "100                                      NaN   \n",
       "101                                      NaN   \n",
       "102                                      NaN   \n",
       "103                                      NaN   \n",
       "104                                      NaN   \n",
       "105                                      0.0   \n",
       "106                                      NaN   \n",
       "107                                      NaN   \n",
       "108                                      NaN   \n",
       "109                                      NaN   \n",
       "110                                      NaN   \n",
       "111                                      NaN   \n",
       "\n",
       "    2b_2b_cut1_total_reads_collapsed_percent  \\\n",
       "0                                        0.0   \n",
       "1                                        NaN   \n",
       "2                                        NaN   \n",
       "3                                        NaN   \n",
       "4                                        NaN   \n",
       "5                                        NaN   \n",
       "6                                        NaN   \n",
       "7                                        0.0   \n",
       "8                                        NaN   \n",
       "9                                        NaN   \n",
       "10                                       NaN   \n",
       "11                                       NaN   \n",
       "12                                       NaN   \n",
       "13                                       NaN   \n",
       "14                                       0.0   \n",
       "15                                       NaN   \n",
       "16                                       NaN   \n",
       "17                                       NaN   \n",
       "18                                       NaN   \n",
       "19                                       NaN   \n",
       "20                                       NaN   \n",
       "21                                       0.0   \n",
       "22                                       NaN   \n",
       "23                                       NaN   \n",
       "24                                       NaN   \n",
       "25                                       NaN   \n",
       "26                                       NaN   \n",
       "27                                       NaN   \n",
       "28                                       0.0   \n",
       "29                                       NaN   \n",
       "..                                       ...   \n",
       "82                                       NaN   \n",
       "83                                       NaN   \n",
       "84                                       0.0   \n",
       "85                                       NaN   \n",
       "86                                       NaN   \n",
       "87                                       NaN   \n",
       "88                                       NaN   \n",
       "89                                       NaN   \n",
       "90                                       NaN   \n",
       "91                                       0.0   \n",
       "92                                       NaN   \n",
       "93                                       NaN   \n",
       "94                                       NaN   \n",
       "95                                       NaN   \n",
       "96                                       NaN   \n",
       "97                                       NaN   \n",
       "98                                       0.0   \n",
       "99                                       NaN   \n",
       "100                                      NaN   \n",
       "101                                      NaN   \n",
       "102                                      NaN   \n",
       "103                                      NaN   \n",
       "104                                      NaN   \n",
       "105                                      0.0   \n",
       "106                                      NaN   \n",
       "107                                      NaN   \n",
       "108                                      NaN   \n",
       "109                                      NaN   \n",
       "110                                      NaN   \n",
       "111                                      NaN   \n",
       "\n",
       "     target_plus_plasmid_total_reads_collapsed_percent  \\\n",
       "0                                                  0.0   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                                  0.0   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                                 0.0   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                                 0.0   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                                 0.0   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "82                                                 NaN   \n",
       "83                                                 NaN   \n",
       "84                                                 0.0   \n",
       "85                                                 NaN   \n",
       "86                                                 NaN   \n",
       "87                                                 NaN   \n",
       "88                                                 NaN   \n",
       "89                                                 NaN   \n",
       "90                                                 NaN   \n",
       "91                                                 0.0   \n",
       "92                                                 NaN   \n",
       "93                                                 NaN   \n",
       "94                                                 NaN   \n",
       "95                                                 NaN   \n",
       "96                                                 NaN   \n",
       "97                                                 NaN   \n",
       "98                                                 0.0   \n",
       "99                                                 NaN   \n",
       "100                                                NaN   \n",
       "101                                                NaN   \n",
       "102                                                NaN   \n",
       "103                                                NaN   \n",
       "104                                                NaN   \n",
       "105                                                0.0   \n",
       "106                                                NaN   \n",
       "107                                                NaN   \n",
       "108                                                NaN   \n",
       "109                                                NaN   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "\n",
       "    plasmid_only_total_reads_collapsed_percent  \\\n",
       "0                                     3.508772   \n",
       "1                                          NaN   \n",
       "2                                          NaN   \n",
       "3                                          NaN   \n",
       "4                                          NaN   \n",
       "5                                          NaN   \n",
       "6                                          NaN   \n",
       "7                                     3.508772   \n",
       "8                                          NaN   \n",
       "9                                          NaN   \n",
       "10                                         NaN   \n",
       "11                                         NaN   \n",
       "12                                         NaN   \n",
       "13                                         NaN   \n",
       "14                                    3.508772   \n",
       "15                                         NaN   \n",
       "16                                         NaN   \n",
       "17                                         NaN   \n",
       "18                                         NaN   \n",
       "19                                         NaN   \n",
       "20                                         NaN   \n",
       "21                                    3.508772   \n",
       "22                                         NaN   \n",
       "23                                         NaN   \n",
       "24                                         NaN   \n",
       "25                                         NaN   \n",
       "26                                         NaN   \n",
       "27                                         NaN   \n",
       "28                                    3.508772   \n",
       "29                                         NaN   \n",
       "..                                         ...   \n",
       "82                                         NaN   \n",
       "83                                         NaN   \n",
       "84                                    3.508772   \n",
       "85                                         NaN   \n",
       "86                                         NaN   \n",
       "87                                         NaN   \n",
       "88                                         NaN   \n",
       "89                                         NaN   \n",
       "90                                         NaN   \n",
       "91                                    3.508772   \n",
       "92                                         NaN   \n",
       "93                                         NaN   \n",
       "94                                         NaN   \n",
       "95                                         NaN   \n",
       "96                                         NaN   \n",
       "97                                         NaN   \n",
       "98                                    3.508772   \n",
       "99                                         NaN   \n",
       "100                                        NaN   \n",
       "101                                        NaN   \n",
       "102                                        NaN   \n",
       "103                                        NaN   \n",
       "104                                        NaN   \n",
       "105                                   3.508772   \n",
       "106                                        NaN   \n",
       "107                                        NaN   \n",
       "108                                        NaN   \n",
       "109                                        NaN   \n",
       "110                                        NaN   \n",
       "111                                        NaN   \n",
       "\n",
       "                                         Type Percent Editing  \n",
       "0                 wt_cut1_total_reads_percent        0.000000  \n",
       "1                 wt_cut1_total_reads_percent             NaN  \n",
       "2                 wt_cut1_total_reads_percent             NaN  \n",
       "3                 wt_cut1_total_reads_percent             NaN  \n",
       "4                 wt_cut1_total_reads_percent             NaN  \n",
       "5                 wt_cut1_total_reads_percent             NaN  \n",
       "6                 wt_cut1_total_reads_percent             NaN  \n",
       "7                 wt_cut2_total_reads_percent        0.000000  \n",
       "8                 wt_cut2_total_reads_percent             NaN  \n",
       "9                 wt_cut2_total_reads_percent             NaN  \n",
       "10                wt_cut2_total_reads_percent             NaN  \n",
       "11                wt_cut2_total_reads_percent             NaN  \n",
       "12                wt_cut2_total_reads_percent             NaN  \n",
       "13                wt_cut2_total_reads_percent             NaN  \n",
       "14    large_deletion_cut1_total_reads_percent        0.000000  \n",
       "15    large_deletion_cut1_total_reads_percent             NaN  \n",
       "16    large_deletion_cut1_total_reads_percent             NaN  \n",
       "17    large_deletion_cut1_total_reads_percent             NaN  \n",
       "18    large_deletion_cut1_total_reads_percent             NaN  \n",
       "19    large_deletion_cut1_total_reads_percent             NaN  \n",
       "20    large_deletion_cut1_total_reads_percent             NaN  \n",
       "21   large_inversion_cut1_total_reads_percent        0.000000  \n",
       "22   large_inversion_cut1_total_reads_percent             NaN  \n",
       "23   large_inversion_cut1_total_reads_percent             NaN  \n",
       "24   large_inversion_cut1_total_reads_percent             NaN  \n",
       "25   large_inversion_cut1_total_reads_percent             NaN  \n",
       "26   large_inversion_cut1_total_reads_percent             NaN  \n",
       "27   large_inversion_cut1_total_reads_percent             NaN  \n",
       "28   large_inversion_cut2_total_reads_percent        0.000000  \n",
       "29   large_inversion_cut2_total_reads_percent             NaN  \n",
       "..                                        ...             ...  \n",
       "82   doner_head_head_cut1_total_reads_percent             NaN  \n",
       "83   doner_head_head_cut1_total_reads_percent             NaN  \n",
       "84             1a_1a_cut1_total_reads_percent        0.000000  \n",
       "85             1a_1a_cut1_total_reads_percent             NaN  \n",
       "86             1a_1a_cut1_total_reads_percent             NaN  \n",
       "87             1a_1a_cut1_total_reads_percent             NaN  \n",
       "88             1a_1a_cut1_total_reads_percent             NaN  \n",
       "89             1a_1a_cut1_total_reads_percent             NaN  \n",
       "90             1a_1a_cut1_total_reads_percent             NaN  \n",
       "91             2b_2b_cut1_total_reads_percent        0.000000  \n",
       "92             2b_2b_cut1_total_reads_percent             NaN  \n",
       "93             2b_2b_cut1_total_reads_percent             NaN  \n",
       "94             2b_2b_cut1_total_reads_percent             NaN  \n",
       "95             2b_2b_cut1_total_reads_percent             NaN  \n",
       "96             2b_2b_cut1_total_reads_percent             NaN  \n",
       "97             2b_2b_cut1_total_reads_percent             NaN  \n",
       "98    target_plus_plasmid_total_reads_percent        0.000000  \n",
       "99    target_plus_plasmid_total_reads_percent             NaN  \n",
       "100   target_plus_plasmid_total_reads_percent             NaN  \n",
       "101   target_plus_plasmid_total_reads_percent             NaN  \n",
       "102   target_plus_plasmid_total_reads_percent             NaN  \n",
       "103   target_plus_plasmid_total_reads_percent             NaN  \n",
       "104   target_plus_plasmid_total_reads_percent             NaN  \n",
       "105          plasmid_only_total_reads_percent        3.508772  \n",
       "106          plasmid_only_total_reads_percent             NaN  \n",
       "107          plasmid_only_total_reads_percent             NaN  \n",
       "108          plasmid_only_total_reads_percent             NaN  \n",
       "109          plasmid_only_total_reads_percent             NaN  \n",
       "110          plasmid_only_total_reads_percent             NaN  \n",
       "111          plasmid_only_total_reads_percent             NaN  \n",
       "\n",
       "[112 rows x 188 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_alignments_junction = pd.concat([result_amplicon_df, result_plasmid_df], axis=1)\n",
    "sample_info_filename = os.path.join(directory, 'sample_info.csv')\n",
    "\n",
    "experiments = pd.read_csv(sample_info_filename)\n",
    "\n",
    "N7 = amplicon_info['index_I1']\n",
    "N5 = amplicon_info['index_I2']\n",
    "\n",
    "results_summary = summarize_results(results_alignments_junction)\n",
    "#this outputs it in a directory above the N501_701 folder as a summary file. Need to label it or it will be overwritten by next tone\n",
    "results_summary_with_experiments = pd.concat([experiments, results_summary], axis=1)\n",
    "\n",
    "results_summary_with_experiments.to_excel(os.path.join(directory, N7 + '_' + N5, N7 + '_' + N5+'_'+'results_summary.xlsx'))\n",
    "\n",
    "results_pivot = melt_results_lam(results_summary_with_experiments)\n",
    "results_pivot.to_excel(os.path.join(directory, N7 + '_' + N5, N7 + '_' + N5+'_'+'results_summary_pivot_no_UMI_collapse.xlsx'))\n",
    "results_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline2\n",
    "#  This is a second pipeline. Different than Uditas but built in a simliar form. \n",
    "## It uses the read distal to the gene specic primer. This read covers the region downstream of the break. After clipping off anything remaing upstream of the break we make a global alignment to look at genomic integrations and translocations. This also quantifies the alignments.\n",
    "\n",
    "\n",
    "### 0. Use correct_prime function in pipeline1 to arrive at reads that were correctly primed.\n",
    "### 1. Trim correctly primed reads to the break site  \n",
    "### 2. Only use a single end of the pair end reads (distal gene specific primer)\n",
    "### 3. Use bowtie2 genome that has the targeting seq in it added\n",
    "### 4. Do global end-to-end alignment\n",
    "### 5. Do a local alignment\n",
    "### 6. process bam file and pull out the high mapQ reads\n",
    "\n",
    "The idea here is to trim the sequence up to the cut on both reads and then do a global alignment end-to-end agasint the human genome.\n",
    "\n",
    "The idea here is to take the files that are 'trimmed' and 'correct priming' checking. Then to do another round of trimming to get down to the cut site. I need to make another folder for 'trimmed_to_cut'\n",
    "\n",
    "For Tn5 setup the 'Read 1' contains the sequence away from the gene specific primer. So it is the read we should concentration on if we don't do paired end.\n",
    "\n",
    "For LAM the 'Read 2' is the the sequence distal to the gene specific primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "this is for pipeline1 LAM mCherryFwd N707 N505, row2 in sheet \n",
      "\n",
      "seq_to_breaksite TGCATCGCATTGTCTGAGTAGGTGTCATTCTATTCTGGGGGGTGGGGTGGGGCAGGACAGCAAGGGGGAGGATTGGGAAGAGAATAGCAGGCATGCTGGGGGGATCCCTGCAGGAGCTCGTCGACCCATGGGGGCCCGCCCCAACTGGGGTAACCTTTGGGCTCCCCGGGCGCGACTAGTGAATTCAGATCTGATATCTCTAGAAGTCCTGGG \n",
      "\n",
      "/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM/10000Reads\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                        LAM_Pol50_bpA\n",
       "Sample                                                                               Polb\n",
       "description                                         MiniSeq_K562_Polb_mCherrFwd_50per_LAM\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                         NaN\n",
       "I7_Index_ID                                                          P7_N707_SBS12nextera\n",
       "index_I1                                                                             N707\n",
       "barcode_I1                                                                       CTCTCTAC\n",
       "I5_Index_ID                                                             P5_UMI_N505_Tn5-A\n",
       "index_I2                                                                             N505\n",
       "barcode_I2                                                                       CTCCTTAC\n",
       "primer_outfwd                                                                         NaN\n",
       "primer                                                                U1-C-bpA_Fwd_nest2 \n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                            42350359\n",
       "end                                                                              42350378\n",
       "strand                                                                                  -\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "genome_plus_targeting                                              hg38_plus_targetvector\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                            pE049_WO_redundant\n",
       "plasmid_sequence                        ttacataacttacggtaaatggcccgcctggctgaccgcccaacga...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 2, dtype: object"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#This is for the sequence to the breaksite for the pipeline of genomic alignment \n",
    "# make sure the sequences are capital\n",
    "\n",
    "print()\n",
    "seq_to_breaksite = [['GGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG', 'this is for pipeline1 LAM- polb rev WT N706 N504 (row 0 in sheet)'],\n",
    "                    ['GGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG','this is for pipeline1 LAM polb rev Targeting 50% N706 N505(row 1 in sheet)'],\n",
    "                    ['TGCATCGCATTGTCTGAGTAGGTGTCATTCTATTCTGGGGGGTGGGGTGGGGCAGGACAGCAAGGGGGAGGATTGGGAAGAGAATAGCAGGCATGCTGGGGGGATCCCTGCAGGAGCTCGTCGACCCATGGGGGCCCGCCCCAACTGGGGTAACCTTTGGGCTCCCCGGGCGCGACTAGTGAATTCAGATCTGATATCTCTAGAAGTCCTGGG','this is for pipeline1 LAM mCherryFwd N707 N505, row2 in sheet'],\n",
    "                    ['CCTCCACGTCACCGCATGTTAGCAGACTTCCTCTGCCCTCAATCTTTAAGAAAAAAAAAAGTCTAACAATGATTTAGGAATGCTTTGAGGACTTAAATGATCTTATTGGAAACATACCAGTCTGCTAAAAGACTAATTTTGTGTGGGTCA', 'this is for mCherryrev LAM  N708 N505 , row3 in sheet']]\n",
    "\n",
    "\n",
    "seq_primer_to_breaksite = seq_to_breaksite[master_number][0]\n",
    "print(seq_to_breaksite[master_number][1],\"\\n\")    \n",
    "print('seq_to_breaksite', seq_primer_to_breaksite, '\\n')\n",
    "\n",
    "\n",
    "#check amplicon info and directory\n",
    "print(directory)\n",
    "amplicon_info\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Need to make an new bowtie2 index file that includes targeting for alignment agains the whole genome so it is all in one sheet together. \n",
    "\n",
    "### Other option would be to align it to the amplicon, extract unaligned files and then align to the genome but seems cleaner this way. \n",
    "#### Ideally every sequence is unique between the targeting vector and genome.\n",
    "\n",
    "\n",
    "1. Build your fastas of interest and label .fa files.\n",
    "    1. You need fasta of hg38 or reference genome. You can pull this from downloaded bowtie indexed sampels and then use the following command to turn the index into a fasta file: bowtie2-inspect hg38 > hg38.fa   \n",
    "    2. Put all the fasta files in the same folder\n",
    "2. index the files with bowtie\n",
    "    1. use the command bowtie2-build -f pe038_mc.fa,hg38.fa -p hg38_plus_targetvector\n",
    "    2. this has the -p to make it take less ram in my case.\n",
    "    3. In this case it adds the hg38 and the minicircle targeting file together\n",
    "    4. I have a Intel® Core™ i7-5500U CPU @ 2.40GHz × 4 with 15.1 GiB ram and it required about 13.8 gigs of ram and 2 hours to do a hg38+small fasta index\n",
    "    5. be sure to pay attention to the name of the new indexed file. \"hg38_plus_targetvector\" in example above. Add it to the sample_info.csv sheet. Under the tab 'genome_plus_targeting'.\n",
    "    6. you can check it indexed correcctly: bowtie2-inspect -s hg38_plus_targetvector\n",
    "    \n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38_plus_targetvector.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "#need a new build\n",
    "assembly_plus_targetvector = amplicon_info['genome_plus_targeting']\n",
    "file_genome_2bit_plus_target = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly_plus_targetvector + '.2bit')\n",
    "print(file_genome_2bit_plus_target)\n",
    "\n",
    "#Needs to have a bowtie build in which the targeting vector is included!!\n",
    "\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "\n",
    "#check the env variables\n",
    "#%env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lam analysis\n"
     ]
    }
   ],
   "source": [
    "#Trim the break up to the cut site\n",
    "trim_fastq_to_break(directory, amplicon_info, seq_primer_to_breaksite, lam_or_tn5='lam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lam analysis\n",
      "sam file deleted\n"
     ]
    }
   ],
   "source": [
    "#this aligns everything after the break to end-to-end \n",
    "align_afterbreak_end_to_end_genome_global(directory, amplicon_info, assembly_plus_targetvector, lam_or_tn5='lam', keep_sam=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lam analysis\n",
      "sam file deleted\n"
     ]
    }
   ],
   "source": [
    "#this aligns everything in the local format wehre it can soft clip the ends\n",
    "align_afterbreaks_genome_local(directory, 1, amplicon_info, assembly_plus_targetvector, lam_or_tn5 ='lam', keep_sam=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
